<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />

<meta name="viewport" content="width=device-width, initial-scale=1" />

<meta name="author" content="Wei Liu" />

<meta name="date" content="2022-10-18" />

<title>PRECAST: Human Breast Cancer Data Analysis</title>


<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>


<style type="text/css">
  code {
    white-space: pre;
  }
  .sourceCode {
    overflow: visible;
  }
</style>
<style type="text/css" data-origin="pandoc">
a.sourceLine { display: inline-block; line-height: 1.25; }
a.sourceLine { pointer-events: none; color: inherit; text-decoration: inherit; }
a.sourceLine:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode { white-space: pre; position: relative; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
code.sourceCode { white-space: pre-wrap; }
a.sourceLine { text-indent: -1em; padding-left: 1em; }
}
pre.numberSource a.sourceLine
  { position: relative; left: -4em; }
pre.numberSource a.sourceLine::before
  { content: attr(title);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; pointer-events: all; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {  }
@media screen {
a.sourceLine::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */

</style>
<script>
// apply pandoc div.sourceCode style to pre.sourceCode instead
(function() {
  var sheets = document.styleSheets;
  for (var i = 0; i < sheets.length; i++) {
    if (sheets[i].ownerNode.dataset["origin"] !== "pandoc") continue;
    try { var rules = sheets[i].cssRules; } catch (e) { continue; }
    for (var j = 0; j < rules.length; j++) {
      var rule = rules[j];
      // check if there is a div.sourceCode rule
      if (rule.type !== rule.STYLE_RULE || rule.selectorText !== "div.sourceCode") continue;
      var style = rule.style.cssText;
      // check if color or background-color is set
      if (rule.style.color === '' && rule.style.backgroundColor === '') continue;
      // replace div.sourceCode by a pre.sourceCode rule
      sheets[i].deleteRule(j);
      sheets[i].insertRule('pre.sourceCode{' + style + '}', j);
    }
  }
})();
</script>




<style type="text/css">body {
background-color: #fff;
margin: 1em auto;
max-width: 700px;
overflow: visible;
padding-left: 2em;
padding-right: 2em;
font-family: "Open Sans", "Helvetica Neue", Helvetica, Arial, sans-serif;
font-size: 14px;
line-height: 1.35;
}
#TOC {
clear: both;
margin: 0 0 10px 10px;
padding: 4px;
width: 400px;
border: 1px solid #CCCCCC;
border-radius: 5px;
background-color: #f6f6f6;
font-size: 13px;
line-height: 1.3;
}
#TOC .toctitle {
font-weight: bold;
font-size: 15px;
margin-left: 5px;
}
#TOC ul {
padding-left: 40px;
margin-left: -1.5em;
margin-top: 5px;
margin-bottom: 5px;
}
#TOC ul ul {
margin-left: -2em;
}
#TOC li {
line-height: 16px;
}
table {
margin: 1em auto;
border-width: 1px;
border-color: #DDDDDD;
border-style: outset;
border-collapse: collapse;
}
table th {
border-width: 2px;
padding: 5px;
border-style: inset;
}
table td {
border-width: 1px;
border-style: inset;
line-height: 18px;
padding: 5px 5px;
}
table, table th, table td {
border-left-style: none;
border-right-style: none;
}
table thead, table tr.even {
background-color: #f7f7f7;
}
p {
margin: 0.5em 0;
}
blockquote {
background-color: #f6f6f6;
padding: 0.25em 0.75em;
}
hr {
border-style: solid;
border: none;
border-top: 1px solid #777;
margin: 28px 0;
}
dl {
margin-left: 0;
}
dl dd {
margin-bottom: 13px;
margin-left: 13px;
}
dl dt {
font-weight: bold;
}
ul {
margin-top: 0;
}
ul li {
list-style: circle outside;
}
ul ul {
margin-bottom: 0;
}
pre, code {
background-color: #f7f7f7;
border-radius: 3px;
color: #333;
white-space: pre-wrap; 
}
pre {
border-radius: 3px;
margin: 5px 0px 10px 0px;
padding: 10px;
}
pre:not([class]) {
background-color: #f7f7f7;
}
code {
font-family: Consolas, Monaco, 'Courier New', monospace;
font-size: 85%;
}
p > code, li > code {
padding: 2px 0px;
}
div.figure {
text-align: center;
}
img {
background-color: #FFFFFF;
padding: 2px;
border: 1px solid #DDDDDD;
border-radius: 3px;
border: 1px solid #CCCCCC;
margin: 0 5px;
}
h1 {
margin-top: 0;
font-size: 35px;
line-height: 40px;
}
h2 {
border-bottom: 4px solid #f7f7f7;
padding-top: 10px;
padding-bottom: 2px;
font-size: 145%;
}
h3 {
border-bottom: 2px solid #f7f7f7;
padding-top: 10px;
font-size: 120%;
}
h4 {
border-bottom: 1px solid #f7f7f7;
margin-left: 8px;
font-size: 105%;
}
h5, h6 {
border-bottom: 1px solid #ccc;
font-size: 105%;
}
a {
color: #0033dd;
text-decoration: none;
}
a:hover {
color: #6666ff; }
a:visited {
color: #800080; }
a:visited:hover {
color: #BB00BB; }
a[href^="http:"] {
text-decoration: underline; }
a[href^="https:"] {
text-decoration: underline; }

code > span.kw { color: #555; font-weight: bold; } 
code > span.dt { color: #902000; } 
code > span.dv { color: #40a070; } 
code > span.bn { color: #d14; } 
code > span.fl { color: #d14; } 
code > span.ch { color: #d14; } 
code > span.st { color: #d14; } 
code > span.co { color: #888888; font-style: italic; } 
code > span.ot { color: #007020; } 
code > span.al { color: #ff0000; font-weight: bold; } 
code > span.fu { color: #900; font-weight: bold; } 
code > span.er { color: #a61717; background-color: #e3d2d2; } 
</style>




</head>

<body>




<h1 class="title toc-ignore">PRECAST: Human Breast Cancer Data Analysis</h1>
<h4 class="author">Wei Liu</h4>
<h4 class="date">2022-10-18</h4>



<p>This vignette introduces the PRECAST workflow for the analysis of integrating multiple spatial transcriptomics datasets. The workflow consists of three steps</p>
<ul>
<li>Independent preprocessing and model setting</li>
<li>Probabilistic embedding, clustering and alignment using PRECAST model</li>
<li>Downstream analysis (i.e.Â visualization of clusters and embeddings, combined differential expression analysis)</li>
</ul>
<p>We demonstrate the use of PRECAST to two sliced human breast cancer Visium data that are <a href="https://github.com/feiyoung/PRECAST/tree/main/vignettes_data">here</a>, which can be downloaded to the current working path by the following command:</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb1-1" title="1">githubURL &lt;-<span class="st"> &quot;https://github.com/feiyoung/PRECAST/blob/main/vignettes_data/bc2.rda?raw=true&quot;</span></a>
<a class="sourceLine" id="cb1-2" title="2"><span class="kw">download.file</span>(githubURL, <span class="st">&quot;bc2.rda&quot;</span>, <span class="dt">mode =</span> <span class="st">&quot;wb&quot;</span>)</a></code></pre></div>
<p>Then load to R</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb2-1" title="1"><span class="kw">load</span>(<span class="st">&quot;bc2.rda&quot;</span>)</a></code></pre></div>
<p>This data is also available at 10X genomics data website:</p>
<ul>
<li>Section1: <a href="https://support.10xgenomics.com/spatial-gene-expression/datasets/1.1.0/V1_Breast_Cancer_Block_A_Section_1" class="uri">https://support.10xgenomics.com/spatial-gene-expression/datasets/1.1.0/V1_Breast_Cancer_Block_A_Section_1</a></li>
<li>Section2: <a href="https://support.10xgenomics.com/spatial-gene-expression/datasets/1.1.0/V1_Breast_Cancer_Block_A_Section_2" class="uri">https://support.10xgenomics.com/spatial-gene-expression/datasets/1.1.0/V1_Breast_Cancer_Block_A_Section_2</a></li>
</ul>
<p>Users require the two folders for each dataset: spatial and filtered_feature_bc_matrix. Then the data can be read by the following commond.</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb3-1" title="1"><span class="co"># library(DR.SC) dir.file &lt;- &#39;Section&#39; ## the folders Section1 and Section2, and each includes</span></a>
<a class="sourceLine" id="cb3-2" title="2"><span class="co"># two folders spatial and filtered_feature_bc_matrix seuList &lt;- list() for (r in 1:2) {</span></a>
<a class="sourceLine" id="cb3-3" title="3"><span class="co"># message(&#39;r = &#39;, r) seuList[[r]] &lt;- read10XVisium(paste0(dir.file, r)) } bc2 &lt;- seuList</span></a></code></pre></div>
<p>The package can be loaded with the command:</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb4-1" title="1"><span class="kw">library</span>(PRECAST)</a>
<a class="sourceLine" id="cb4-2" title="2"><span class="kw">library</span>(Seurat)</a></code></pre></div>
<div id="fit-precast-using-this-data" class="section level2">
<h2>Fit PRECAST using this data</h2>
<p>View human breast cancer Visium data from <code>DataPRECAST</code></p>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb5-1" title="1">bc2  <span class="co">## a list including two Seurat object</span></a></code></pre></div>
<p>Check the content in <code>bc2</code></p>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb6-1" title="1"><span class="kw">head</span>(bc2[[<span class="dv">1</span>]])</a></code></pre></div>
<div id="prepare-the-precastobject." class="section level3">
<h3>Prepare the PRECASTObject.</h3>
<p>Human breast cancer data have been pre-processed and saved as the <code>PRECASTObj</code> format.</p>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb7-1" title="1"><span class="co">## Create PRECASTObject.</span></a>
<a class="sourceLine" id="cb7-2" title="2"><span class="kw">set.seed</span>(<span class="dv">2022</span>)</a>
<a class="sourceLine" id="cb7-3" title="3">PRECASTObj &lt;-<span class="st"> </span><span class="kw">CreatePRECASTObject</span>(bc2)</a></code></pre></div>
</div>
<div id="add-the-model-setting" class="section level3">
<h3>Add the model setting</h3>
<div class="sourceCode" id="cb8"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb8-1" title="1"><span class="co">## check the number of genes/features after filtering step</span></a>
<a class="sourceLine" id="cb8-2" title="2">PRECASTObj<span class="op">@</span>seulist</a>
<a class="sourceLine" id="cb8-3" title="3"></a>
<a class="sourceLine" id="cb8-4" title="4"><span class="co">## Add adjacency matrix list for a PRECASTObj object to prepare for PRECAST model fitting.</span></a>
<a class="sourceLine" id="cb8-5" title="5">PRECASTObj &lt;-<span class="st"> </span><span class="kw">AddAdjList</span>(PRECASTObj, <span class="dt">platform =</span> <span class="st">&quot;Visium&quot;</span>)</a>
<a class="sourceLine" id="cb8-6" title="6"></a>
<a class="sourceLine" id="cb8-7" title="7"><span class="co">## Add a model setting in advance for a PRECASTObj object. verbose =TRUE helps outputing the</span></a>
<a class="sourceLine" id="cb8-8" title="8"><span class="co">## information in the algorithm.</span></a>
<a class="sourceLine" id="cb8-9" title="9">PRECASTObj &lt;-<span class="st"> </span><span class="kw">AddParSetting</span>(PRECASTObj, <span class="dt">Sigma_equal =</span> <span class="ot">FALSE</span>, <span class="dt">verbose =</span> <span class="ot">TRUE</span>, <span class="dt">int.model =</span> <span class="ot">NULL</span>)</a></code></pre></div>
</div>
<div id="fit-precast" class="section level3">
<h3>Fit PRECAST</h3>
<p>For function <code>PRECAST</code>, users can specify the number of clusters <span class="math inline">\(K\)</span> or set <code>K</code> to be an integer vector by using modified BIC(MBIC) to determine <span class="math inline">\(K\)</span>. First, we try using user-specified number of clusters. For convenience, we give the selected number of clusters by MBIC (K=14).</p>
<div class="sourceCode" id="cb9"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb9-1" title="1"><span class="co">### Given K</span></a>
<a class="sourceLine" id="cb9-2" title="2">PRECASTObj &lt;-<span class="st"> </span><span class="kw">PRECAST</span>(PRECASTObj, <span class="dt">K =</span> <span class="dv">14</span>)</a></code></pre></div>
<p>Select a best model</p>
<div class="sourceCode" id="cb10"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb10-1" title="1"><span class="co">## backup the fitting results in resList</span></a>
<a class="sourceLine" id="cb10-2" title="2">resList &lt;-<span class="st"> </span>PRECASTObj<span class="op">@</span>resList</a>
<a class="sourceLine" id="cb10-3" title="3">PRECASTObj &lt;-<span class="st"> </span><span class="kw">selectModel</span>(PRECASTObj)</a></code></pre></div>
<p>Integrate the two samples by the function <code>IntegrateSpaData</code>.</p>
<div class="sourceCode" id="cb11"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb11-1" title="1">seuInt &lt;-<span class="st"> </span><span class="kw">IntegrateSpaData</span>(PRECASTObj, <span class="dt">species =</span> <span class="st">&quot;Human&quot;</span>)</a>
<a class="sourceLine" id="cb11-2" title="2">seuInt</a>
<a class="sourceLine" id="cb11-3" title="3"><span class="co">## The low-dimensional embeddings obtained by PRECAST are saved in PRECAST reduction slot.</span></a></code></pre></div>
<p>Show the spatial scatter plot for clusters</p>
<div class="sourceCode" id="cb12"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb12-1" title="1">cols_cluster &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;#FD7446&quot;</span>, <span class="st">&quot;#709AE1&quot;</span>, <span class="st">&quot;#31A354&quot;</span>, <span class="st">&quot;#9EDAE5&quot;</span>, <span class="st">&quot;#DE9ED6&quot;</span>, <span class="st">&quot;#BCBD22&quot;</span>, <span class="st">&quot;#CE6DBD&quot;</span>, <span class="st">&quot;#DADAEB&quot;</span>,</a>
<a class="sourceLine" id="cb12-2" title="2">    <span class="st">&quot;yellow&quot;</span>, <span class="st">&quot;#FF9896&quot;</span>, <span class="st">&quot;#91D1C2&quot;</span>, <span class="st">&quot;#C7E9C0&quot;</span>, <span class="st">&quot;#6B6ECF&quot;</span>, <span class="st">&quot;#7B4173&quot;</span>)</a>
<a class="sourceLine" id="cb12-3" title="3">p12 &lt;-<span class="st"> </span><span class="kw">SpaPlot</span>(seuInt, <span class="dt">batch =</span> <span class="ot">NULL</span>, <span class="dt">point_size =</span> <span class="dv">1</span>, <span class="dt">cols =</span> cols_cluster, <span class="dt">combine =</span> <span class="ot">TRUE</span>)</a>
<a class="sourceLine" id="cb12-4" title="4">p12</a>
<a class="sourceLine" id="cb12-5" title="5"><span class="co"># users can plot each sample by setting combine=FALSE</span></a></code></pre></div>
<p>Show the spatial UMAP/tNSE RGB plot to illustrate the performance in extracting features.</p>
<div class="sourceCode" id="cb13"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb13-1" title="1">seuInt &lt;-<span class="st"> </span><span class="kw">AddUMAP</span>(seuInt)</a>
<a class="sourceLine" id="cb13-2" title="2">p13 &lt;-<span class="st"> </span><span class="kw">SpaPlot</span>(seuInt, <span class="dt">batch =</span> <span class="ot">NULL</span>, <span class="dt">item =</span> <span class="st">&quot;RGB_UMAP&quot;</span>, <span class="dt">point_size =</span> <span class="dv">2</span>, <span class="dt">combine =</span> <span class="ot">TRUE</span>, <span class="dt">text_size =</span> <span class="dv">15</span>)</a>
<a class="sourceLine" id="cb13-3" title="3">p13</a>
<a class="sourceLine" id="cb13-4" title="4"><span class="co"># seuInt &lt;- AddTSNE(seuInt) SpaPlot(seuInt, batch=NULL,item=&#39;RGB_TSNE&#39;,point_size=2,</span></a>
<a class="sourceLine" id="cb13-5" title="5"><span class="co"># combine=T, text_size=15)</span></a></code></pre></div>
<p>Show the tSNE plot based on the extracted features from PRECAST to check the performance of integration.</p>
<div class="sourceCode" id="cb14"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb14-1" title="1">seuInt &lt;-<span class="st"> </span><span class="kw">AddTSNE</span>(seuInt, <span class="dt">n_comp =</span> <span class="dv">2</span>)</a>
<a class="sourceLine" id="cb14-2" title="2"><span class="kw">library</span>(patchwork)</a>
<a class="sourceLine" id="cb14-3" title="3">p1 &lt;-<span class="st"> </span><span class="kw">dimPlot</span>(seuInt, <span class="dt">item =</span> <span class="st">&quot;cluster&quot;</span>, <span class="dt">point_size =</span> <span class="fl">0.5</span>, <span class="dt">font_family =</span> <span class="st">&quot;serif&quot;</span>, <span class="dt">cols =</span> cols_cluster)  <span class="co"># Times New Roman</span></a>
<a class="sourceLine" id="cb14-4" title="4">p2 &lt;-<span class="st"> </span><span class="kw">dimPlot</span>(seuInt, <span class="dt">item =</span> <span class="st">&quot;batch&quot;</span>, <span class="dt">point_size =</span> <span class="fl">0.5</span>, <span class="dt">font_family =</span> <span class="st">&quot;serif&quot;</span>)</a>
<a class="sourceLine" id="cb14-5" title="5">p1 <span class="op">+</span><span class="st"> </span>p2</a></code></pre></div>
<p>Combined differential expression analysis</p>
<div class="sourceCode" id="cb15"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb15-1" title="1"><span class="kw">library</span>(Seurat)</a>
<a class="sourceLine" id="cb15-2" title="2">dat_deg &lt;-<span class="st"> </span><span class="kw">FindAllMarkers</span>(seuInt)</a>
<a class="sourceLine" id="cb15-3" title="3"><span class="kw">library</span>(dplyr)</a>
<a class="sourceLine" id="cb15-4" title="4">n &lt;-<span class="st"> </span><span class="dv">10</span></a>
<a class="sourceLine" id="cb15-5" title="5">dat_deg <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb15-6" title="6"><span class="st">    </span><span class="kw">group_by</span>(cluster) <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb15-7" title="7"><span class="st">    </span><span class="kw">top_n</span>(<span class="dt">n =</span> n, <span class="dt">wt =</span> avg_log2FC) -&gt;<span class="st"> </span>top10</a>
<a class="sourceLine" id="cb15-8" title="8"></a>
<a class="sourceLine" id="cb15-9" title="9">seuInt &lt;-<span class="st"> </span><span class="kw">ScaleData</span>(seuInt)</a>
<a class="sourceLine" id="cb15-10" title="10">seus &lt;-<span class="st"> </span><span class="kw">subset</span>(seuInt, <span class="dt">downsample =</span> <span class="dv">400</span>)</a></code></pre></div>
<p>Plot DE genesâ heatmap for each spatial domains identified by PRECAST.</p>
<div class="sourceCode" id="cb16"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb16-1" title="1">color_id &lt;-<span class="st"> </span><span class="kw">as.numeric</span>(<span class="kw">levels</span>(<span class="kw">Idents</span>(seus)))</a>
<a class="sourceLine" id="cb16-2" title="2"><span class="kw">library</span>(ggplot2)</a>
<a class="sourceLine" id="cb16-3" title="3"><span class="co">## HeatMap</span></a>
<a class="sourceLine" id="cb16-4" title="4">p1 &lt;-<span class="st"> </span><span class="kw">doHeatmap</span>(seus, <span class="dt">features =</span> top10<span class="op">$</span>gene, <span class="dt">cell_label =</span> <span class="st">&quot;Domain&quot;</span>, <span class="dt">grp_label =</span> F, <span class="dt">grp_color =</span> cols_cluster[color_id],</a>
<a class="sourceLine" id="cb16-5" title="5">    <span class="dt">pt_size =</span> <span class="dv">6</span>, <span class="dt">slot =</span> <span class="st">&quot;scale.data&quot;</span>) <span class="op">+</span><span class="st"> </span><span class="kw">theme</span>(<span class="dt">legend.text =</span> <span class="kw">element_text</span>(<span class="dt">size =</span> <span class="dv">10</span>), <span class="dt">legend.title =</span> <span class="kw">element_text</span>(<span class="dt">size =</span> <span class="dv">13</span>,</a>
<a class="sourceLine" id="cb16-6" title="6">    <span class="dt">face =</span> <span class="st">&quot;bold&quot;</span>), <span class="dt">axis.text.y =</span> <span class="kw">element_text</span>(<span class="dt">size =</span> <span class="dv">5</span>, <span class="dt">face =</span> <span class="st">&quot;italic&quot;</span>, <span class="dt">family =</span> <span class="st">&quot;serif&quot;</span>))</a>
<a class="sourceLine" id="cb16-7" title="7">p1</a></code></pre></div>
</div>
</div>
<div id="session-information" class="section level2">
<h2>Session information</h2>
<div class="sourceCode" id="cb17"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb17-1" title="1"><span class="kw">sessionInfo</span>()</a></code></pre></div>
</div>



<!-- code folding -->


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
