<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />

<meta name="viewport" content="width=device-width, initial-scale=1" />

<meta name="author" content="Wei Liu" />

<meta name="date" content="2025-10-01" />

<title>PRECAST: Human Breast Cancer Data Analysis</title>

<script>// Pandoc 2.9 adds attributes on both header and div. We remove the former (to
// be compatible with the behavior of Pandoc < 2.8).
document.addEventListener('DOMContentLoaded', function(e) {
  var hs = document.querySelectorAll("div.section[class*='level'] > :first-child");
  var i, h, a;
  for (i = 0; i < hs.length; i++) {
    h = hs[i];
    if (!/^h[1-6]$/i.test(h.tagName)) continue;  // it should be a header h1-h6
    a = h.attributes;
    while (a.length > 0) h.removeAttribute(a[0].name);
  }
});
</script>

<style type="text/css">
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
span.underline{text-decoration: underline;}
div.column{display: inline-block; vertical-align: top; width: 50%;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
</style>



<style type="text/css">
code {
white-space: pre;
}
.sourceCode {
overflow: visible;
}
</style>
<style type="text/css" data-origin="pandoc">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
{ counter-reset: source-line 0; }
pre.numberSource code > span
{ position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
{ content: counter(source-line);
position: relative; left: -1em; text-align: right; vertical-align: baseline;
border: none; display: inline-block;
-webkit-touch-callout: none; -webkit-user-select: none;
-khtml-user-select: none; -moz-user-select: none;
-ms-user-select: none; user-select: none;
padding: 0 4px; width: 4em;
color: #aaaaaa;
}
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa; padding-left: 4px; }
div.sourceCode
{ }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } 
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } 
code span.at { color: #7d9029; } 
code span.bn { color: #40a070; } 
code span.bu { color: #008000; } 
code span.cf { color: #007020; font-weight: bold; } 
code span.ch { color: #4070a0; } 
code span.cn { color: #880000; } 
code span.co { color: #60a0b0; font-style: italic; } 
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } 
code span.do { color: #ba2121; font-style: italic; } 
code span.dt { color: #902000; } 
code span.dv { color: #40a070; } 
code span.er { color: #ff0000; font-weight: bold; } 
code span.ex { } 
code span.fl { color: #40a070; } 
code span.fu { color: #06287e; } 
code span.im { color: #008000; font-weight: bold; } 
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } 
code span.kw { color: #007020; font-weight: bold; } 
code span.op { color: #666666; } 
code span.ot { color: #007020; } 
code span.pp { color: #bc7a00; } 
code span.sc { color: #4070a0; } 
code span.ss { color: #bb6688; } 
code span.st { color: #4070a0; } 
code span.va { color: #19177c; } 
code span.vs { color: #4070a0; } 
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } 
</style>
<script>
// apply pandoc div.sourceCode style to pre.sourceCode instead
(function() {
  var sheets = document.styleSheets;
  for (var i = 0; i < sheets.length; i++) {
    if (sheets[i].ownerNode.dataset["origin"] !== "pandoc") continue;
    try { var rules = sheets[i].cssRules; } catch (e) { continue; }
    var j = 0;
    while (j < rules.length) {
      var rule = rules[j];
      // check if there is a div.sourceCode rule
      if (rule.type !== rule.STYLE_RULE || rule.selectorText !== "div.sourceCode") {
        j++;
        continue;
      }
      var style = rule.style.cssText;
      // check if color or background-color is set
      if (rule.style.color === '' && rule.style.backgroundColor === '') {
        j++;
        continue;
      }
      // replace div.sourceCode by a pre.sourceCode rule
      sheets[i].deleteRule(j);
      sheets[i].insertRule('pre.sourceCode{' + style + '}', j);
    }
  }
})();
</script>




<style type="text/css">body {
background-color: #fff;
margin: 1em auto;
max-width: 700px;
overflow: visible;
padding-left: 2em;
padding-right: 2em;
font-family: "Open Sans", "Helvetica Neue", Helvetica, Arial, sans-serif;
font-size: 14px;
line-height: 1.35;
}
#TOC {
clear: both;
margin: 0 0 10px 10px;
padding: 4px;
width: 400px;
border: 1px solid #CCCCCC;
border-radius: 5px;
background-color: #f6f6f6;
font-size: 13px;
line-height: 1.3;
}
#TOC .toctitle {
font-weight: bold;
font-size: 15px;
margin-left: 5px;
}
#TOC ul {
padding-left: 40px;
margin-left: -1.5em;
margin-top: 5px;
margin-bottom: 5px;
}
#TOC ul ul {
margin-left: -2em;
}
#TOC li {
line-height: 16px;
}
table {
margin: 1em auto;
border-width: 1px;
border-color: #DDDDDD;
border-style: outset;
border-collapse: collapse;
}
table th {
border-width: 2px;
padding: 5px;
border-style: inset;
}
table td {
border-width: 1px;
border-style: inset;
line-height: 18px;
padding: 5px 5px;
}
table, table th, table td {
border-left-style: none;
border-right-style: none;
}
table thead, table tr.even {
background-color: #f7f7f7;
}
p {
margin: 0.5em 0;
}
blockquote {
background-color: #f6f6f6;
padding: 0.25em 0.75em;
}
hr {
border-style: solid;
border: none;
border-top: 1px solid #777;
margin: 28px 0;
}
dl {
margin-left: 0;
}
dl dd {
margin-bottom: 13px;
margin-left: 13px;
}
dl dt {
font-weight: bold;
}
ul {
margin-top: 0;
}
ul li {
list-style: circle outside;
}
ul ul {
margin-bottom: 0;
}
pre, code {
background-color: #f7f7f7;
border-radius: 3px;
color: #333;
white-space: pre-wrap; 
}
pre {
border-radius: 3px;
margin: 5px 0px 10px 0px;
padding: 10px;
}
pre:not([class]) {
background-color: #f7f7f7;
}
code {
font-family: Consolas, Monaco, 'Courier New', monospace;
font-size: 85%;
}
p > code, li > code {
padding: 2px 0px;
}
div.figure {
text-align: center;
}
img {
background-color: #FFFFFF;
padding: 2px;
border: 1px solid #DDDDDD;
border-radius: 3px;
border: 1px solid #CCCCCC;
margin: 0 5px;
}
h1 {
margin-top: 0;
font-size: 35px;
line-height: 40px;
}
h2 {
border-bottom: 4px solid #f7f7f7;
padding-top: 10px;
padding-bottom: 2px;
font-size: 145%;
}
h3 {
border-bottom: 2px solid #f7f7f7;
padding-top: 10px;
font-size: 120%;
}
h4 {
border-bottom: 1px solid #f7f7f7;
margin-left: 8px;
font-size: 105%;
}
h5, h6 {
border-bottom: 1px solid #ccc;
font-size: 105%;
}
a {
color: #0033dd;
text-decoration: none;
}
a:hover {
color: #6666ff; }
a:visited {
color: #800080; }
a:visited:hover {
color: #BB00BB; }
a[href^="http:"] {
text-decoration: underline; }
a[href^="https:"] {
text-decoration: underline; }

code > span.kw { color: #555; font-weight: bold; } 
code > span.dt { color: #902000; } 
code > span.dv { color: #40a070; } 
code > span.bn { color: #d14; } 
code > span.fl { color: #d14; } 
code > span.ch { color: #d14; } 
code > span.st { color: #d14; } 
code > span.co { color: #888888; font-style: italic; } 
code > span.ot { color: #007020; } 
code > span.al { color: #ff0000; font-weight: bold; } 
code > span.fu { color: #900; font-weight: bold; } 
code > span.er { color: #a61717; background-color: #e3d2d2; } 
</style>




</head>

<body>




<h1 class="title toc-ignore">PRECAST: Human Breast Cancer Data
Analysis</h1>
<h4 class="author">Wei Liu</h4>
<h4 class="date">2025-10-01</h4>



<p>This vignette introduces the PRECAST workflow for the analysis of
integrating multiple spatial transcriptomics datasets. The workflow
consists of three steps</p>
<ul>
<li>Independent preprocessing and model setting</li>
<li>Probabilistic embedding, clustering and alignment using PRECAST
model</li>
<li>Downstream analysis (i.e.Â visualization of clusters and embeddings,
combined differential expression analysis)</li>
</ul>
<p>We demonstrate the use of PRECAST to two sliced human breast cancer
Visium data that are <a href="https://github.com/feiyoung/PRECAST/tree/main/vignettes_data">here</a>,
which can be downloaded to the current working path by the following
command:</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" tabindex="-1"></a>githubURL <span class="ot">&lt;-</span> <span class="st">&quot;https://github.com/feiyoung/PRECAST/blob/main/vignettes_data/bc2.rda?raw=true&quot;</span></span>
<span id="cb1-2"><a href="#cb1-2" tabindex="-1"></a><span class="fu">download.file</span>(githubURL, <span class="st">&quot;bc2.rda&quot;</span>, <span class="at">mode =</span> <span class="st">&quot;wb&quot;</span>)</span></code></pre></div>
<p>Then load to R</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" tabindex="-1"></a><span class="fu">load</span>(<span class="st">&quot;bc2.rda&quot;</span>)</span></code></pre></div>
<details>
<summary>
<strong>Download data from 10X: another method to access data</strong>
</summary>
<p>This data is also available at 10X genomics data website:</p>
<ul>
<li>Section1: <a href="https://support.10xgenomics.com/spatial-gene-expression/datasets/1.1.0/V1_Breast_Cancer_Block_A_Section_1" class="uri">https://support.10xgenomics.com/spatial-gene-expression/datasets/1.1.0/V1_Breast_Cancer_Block_A_Section_1</a></li>
<li>Section2: <a href="https://support.10xgenomics.com/spatial-gene-expression/datasets/1.1.0/V1_Breast_Cancer_Block_A_Section_2" class="uri">https://support.10xgenomics.com/spatial-gene-expression/datasets/1.1.0/V1_Breast_Cancer_Block_A_Section_2</a></li>
</ul>
<p>Users require the two folders for each dataset: spatial and
filtered_feature_bc_matrix. Then the data can be read by the following
commond.</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" tabindex="-1"></a>dir.file <span class="ot">&lt;-</span> <span class="st">&quot;Section&quot;</span>  <span class="do">## the folders Section1 and Section2, and each includes two folders spatial and filtered_feature_bc_matrix</span></span>
<span id="cb3-2"><a href="#cb3-2" tabindex="-1"></a>seuList <span class="ot">&lt;-</span> <span class="fu">list</span>()</span>
<span id="cb3-3"><a href="#cb3-3" tabindex="-1"></a><span class="cf">for</span> (r <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">2</span>) {</span>
<span id="cb3-4"><a href="#cb3-4" tabindex="-1"></a>    <span class="fu">message</span>(<span class="st">&quot;r = &quot;</span>, r)</span>
<span id="cb3-5"><a href="#cb3-5" tabindex="-1"></a>    seuList[[r]] <span class="ot">&lt;-</span> DR.SC<span class="sc">::</span><span class="fu">read10XVisium</span>(<span class="fu">paste0</span>(dir.file, r))</span>
<span id="cb3-6"><a href="#cb3-6" tabindex="-1"></a>}</span>
<span id="cb3-7"><a href="#cb3-7" tabindex="-1"></a>bc2 <span class="ot">&lt;-</span> seuList</span></code></pre></div>
</details>
<p>The package can be loaded with the command:</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" tabindex="-1"></a><span class="fu">library</span>(PRECAST)</span>
<span id="cb4-2"><a href="#cb4-2" tabindex="-1"></a><span class="fu">library</span>(Seurat)</span></code></pre></div>
<p>View human breast cancer Visium data from
<code>DataPRECAST</code></p>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" tabindex="-1"></a>bc2  <span class="do">## a list including two Seurat object</span></span></code></pre></div>
<p>Check the content in <code>bc2</code></p>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" tabindex="-1"></a><span class="fu">head</span>(bc2[[<span class="dv">1</span>]])</span></code></pre></div>
<div id="create-a-precastobject-object" class="section level2">
<h2>Create a PRECASTObject object</h2>
<p>We show how to create a PRECASTObject object step by step. First, we
create a Seurat list object using the count matrix and meta data of each
data batch. Although <code>bc2</code> is a prepared Seurat list object,
we re-create it to show the details of the Seurat list object. At the
same time, check the meta data that must include the spatial coordinates
named ârowâ and âcolâ, respectively. If the names are not, they are
required to rename them.</p>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" tabindex="-1"></a><span class="do">## Get the gene-by-spot read count matrices countList &lt;- lapply(bc2, function(x)</span></span>
<span id="cb7-2"><a href="#cb7-2" tabindex="-1"></a><span class="do">## x[[&#39;RNA&#39;]]@counts)</span></span>
<span id="cb7-3"><a href="#cb7-3" tabindex="-1"></a>countList <span class="ot">&lt;-</span> <span class="fu">lapply</span>(bc2, <span class="cf">function</span>(x) {</span>
<span id="cb7-4"><a href="#cb7-4" tabindex="-1"></a>    assay <span class="ot">&lt;-</span> <span class="fu">DefaultAssay</span>(x)</span>
<span id="cb7-5"><a href="#cb7-5" tabindex="-1"></a>    <span class="fu">GetAssayData</span>(x, <span class="at">assay =</span> assay, <span class="at">slot =</span> <span class="st">&quot;counts&quot;</span>)</span>
<span id="cb7-6"><a href="#cb7-6" tabindex="-1"></a></span>
<span id="cb7-7"><a href="#cb7-7" tabindex="-1"></a>})</span>
<span id="cb7-8"><a href="#cb7-8" tabindex="-1"></a></span>
<span id="cb7-9"><a href="#cb7-9" tabindex="-1"></a>M <span class="ot">&lt;-</span> <span class="fu">length</span>(countList)</span>
<span id="cb7-10"><a href="#cb7-10" tabindex="-1"></a><span class="do">## Get the meta data of each spot for each data batch</span></span>
<span id="cb7-11"><a href="#cb7-11" tabindex="-1"></a>metadataList <span class="ot">&lt;-</span> <span class="fu">lapply</span>(bc2, <span class="cf">function</span>(x) x<span class="sc">@</span>meta.data)</span>
<span id="cb7-12"><a href="#cb7-12" tabindex="-1"></a></span>
<span id="cb7-13"><a href="#cb7-13" tabindex="-1"></a><span class="cf">for</span> (r <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span>M) {</span>
<span id="cb7-14"><a href="#cb7-14" tabindex="-1"></a>    meta_data <span class="ot">&lt;-</span> metadataList[[r]]</span>
<span id="cb7-15"><a href="#cb7-15" tabindex="-1"></a>    <span class="fu">all</span>(<span class="fu">c</span>(<span class="st">&quot;row&quot;</span>, <span class="st">&quot;col&quot;</span>) <span class="sc">%in%</span> <span class="fu">colnames</span>(meta_data))  <span class="do">## the names are correct!</span></span>
<span id="cb7-16"><a href="#cb7-16" tabindex="-1"></a>    <span class="fu">head</span>(meta_data[, <span class="fu">c</span>(<span class="st">&quot;row&quot;</span>, <span class="st">&quot;col&quot;</span>)])</span>
<span id="cb7-17"><a href="#cb7-17" tabindex="-1"></a>}</span>
<span id="cb7-18"><a href="#cb7-18" tabindex="-1"></a></span>
<span id="cb7-19"><a href="#cb7-19" tabindex="-1"></a></span>
<span id="cb7-20"><a href="#cb7-20" tabindex="-1"></a><span class="do">## ensure the row.names of metadata in metaList are the same as that of colnames count matrix</span></span>
<span id="cb7-21"><a href="#cb7-21" tabindex="-1"></a><span class="do">## in countList</span></span>
<span id="cb7-22"><a href="#cb7-22" tabindex="-1"></a></span>
<span id="cb7-23"><a href="#cb7-23" tabindex="-1"></a><span class="cf">for</span> (r <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span>M) {</span>
<span id="cb7-24"><a href="#cb7-24" tabindex="-1"></a>    <span class="fu">row.names</span>(metadataList[[r]]) <span class="ot">&lt;-</span> <span class="fu">colnames</span>(countList[[r]])</span>
<span id="cb7-25"><a href="#cb7-25" tabindex="-1"></a>}</span>
<span id="cb7-26"><a href="#cb7-26" tabindex="-1"></a></span>
<span id="cb7-27"><a href="#cb7-27" tabindex="-1"></a></span>
<span id="cb7-28"><a href="#cb7-28" tabindex="-1"></a><span class="do">## Create the Seurat list object</span></span>
<span id="cb7-29"><a href="#cb7-29" tabindex="-1"></a></span>
<span id="cb7-30"><a href="#cb7-30" tabindex="-1"></a>seuList <span class="ot">&lt;-</span> <span class="fu">list</span>()</span>
<span id="cb7-31"><a href="#cb7-31" tabindex="-1"></a><span class="cf">for</span> (r <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span>M) {</span>
<span id="cb7-32"><a href="#cb7-32" tabindex="-1"></a>    seuList[[r]] <span class="ot">&lt;-</span> <span class="fu">CreateSeuratObject</span>(<span class="at">counts =</span> countList[[r]], <span class="at">meta.data =</span> metadataList[[r]], <span class="at">project =</span> <span class="st">&quot;BreastCancerPRECAST&quot;</span>)</span>
<span id="cb7-33"><a href="#cb7-33" tabindex="-1"></a>}</span>
<span id="cb7-34"><a href="#cb7-34" tabindex="-1"></a></span>
<span id="cb7-35"><a href="#cb7-35" tabindex="-1"></a>bc2 <span class="ot">&lt;-</span> seuList</span>
<span id="cb7-36"><a href="#cb7-36" tabindex="-1"></a><span class="fu">rm</span>(seuList)</span>
<span id="cb7-37"><a href="#cb7-37" tabindex="-1"></a><span class="fu">head</span>(meta_data[, <span class="fu">c</span>(<span class="st">&quot;row&quot;</span>, <span class="st">&quot;col&quot;</span>)])</span></code></pre></div>
<div id="prepare-the-precastobject-with-preprocessing-step." class="section level3">
<h3>Prepare the PRECASTObject with preprocessing step.</h3>
<p>Next, we use <code>CreatePRECASTObject()</code> to create a
PRECASTObject based on the Seurat list object <code>bc2</code>. This
function will do three things:</p>
<ul>
<li><ol style="list-style-type: decimal">
<li>Filter low-quality spots and genes, controlled by the arguments
<code>premin.features</code> and <code>premin.spots</code>,
respectively; the spots are retained in raw data (bc2) with at least
premin.features number of nonzero-count features (genes), and the genes
are retained in raw data (bc2) with at least <code>premin.spots</code>
number of spots. To ease presentation, we denote the filtered Seurat
list object as bc2_filter1.</li>
</ol></li>
<li><ol start="2" style="list-style-type: decimal">
<li>Select the top 2,000 variable genes (by setting
<code>gene.number=2000</code>) for each data batch using
<code>FindSVGs()</code> function in <code>DR.SC</code> package for
spatially variable genes or <code>FindVariableFeatures()</code> function
in <code>Seurat</code> package for highly variable genes. Next, we
prioritized genes based on the number of times they were selected as
variable genes in all samples and chose the top 2,000 genes. Then denote
the Seurat list object as bc2_filter2, where only 2,000 genes are
retained.</li>
</ol></li>
<li><ol start="3" style="list-style-type: decimal">
<li>Conduct strict quality control for bc2_filter2 by filtering spots
and genes, controlled by the arguments <code>postmin.features</code> and
<code>postmin.spots</code>, respectively; the spots are retained with at
least <code>post.features</code> nonzero counts across genes; the
features (genes) are retained with at least <code>postmin.spots</code>
number of nonzero-count spots. Usually, no genes are filltered because
these genes are variable genes.</li>
</ol></li>
</ul>
<p>If the argument <code>customGenelist</code> is not <code>NULL</code>,
then this function only does (3) not (1) and (2). User can retain the
raw seurat list object by setting
<code>rawData.preserve = TRUE</code>.</p>
<div class="sourceCode" id="cb8"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" tabindex="-1"></a><span class="do">## Create PRECASTObject.</span></span>
<span id="cb8-2"><a href="#cb8-2" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">2022</span>)</span>
<span id="cb8-3"><a href="#cb8-3" tabindex="-1"></a>PRECASTObj <span class="ot">&lt;-</span> <span class="fu">CreatePRECASTObject</span>(bc2, <span class="at">project =</span> <span class="st">&quot;BC2&quot;</span>, <span class="at">gene.number =</span> <span class="dv">2000</span>, <span class="at">selectGenesMethod =</span> <span class="st">&quot;SPARK-X&quot;</span>,</span>
<span id="cb8-4"><a href="#cb8-4" tabindex="-1"></a>    <span class="at">premin.spots =</span> <span class="dv">20</span>, <span class="at">premin.features =</span> <span class="dv">20</span>, <span class="at">postmin.spots =</span> <span class="dv">1</span>, <span class="at">postmin.features =</span> <span class="dv">10</span>)</span>
<span id="cb8-5"><a href="#cb8-5" tabindex="-1"></a></span>
<span id="cb8-6"><a href="#cb8-6" tabindex="-1"></a><span class="do">## User can retain the raw seuList by the following commond.  PRECASTObj &lt;-</span></span>
<span id="cb8-7"><a href="#cb8-7" tabindex="-1"></a><span class="do">## CreatePRECASTObject(seuList, customGenelist=row.names(seuList[[1]]), rawData.preserve =</span></span>
<span id="cb8-8"><a href="#cb8-8" tabindex="-1"></a><span class="do">## TRUE)</span></span></code></pre></div>
</div>
<div id="add-the-model-setting" class="section level3">
<h3>Add the model setting</h3>
<p>Add adjacency matrix list and parameter setting of PRECAST. More
model setting parameters can be found in .</p>
<div class="sourceCode" id="cb9"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" tabindex="-1"></a><span class="do">## check the number of genes/features after filtering step</span></span>
<span id="cb9-2"><a href="#cb9-2" tabindex="-1"></a>PRECASTObj<span class="sc">@</span>seulist</span>
<span id="cb9-3"><a href="#cb9-3" tabindex="-1"></a></span>
<span id="cb9-4"><a href="#cb9-4" tabindex="-1"></a><span class="do">## seuList is null since the default value `rawData.preserve` is FALSE.</span></span>
<span id="cb9-5"><a href="#cb9-5" tabindex="-1"></a>PRECASTObj<span class="sc">@</span>seuList</span>
<span id="cb9-6"><a href="#cb9-6" tabindex="-1"></a></span>
<span id="cb9-7"><a href="#cb9-7" tabindex="-1"></a><span class="do">## Add adjacency matrix list for a PRECASTObj object to prepare for PRECAST model fitting.</span></span>
<span id="cb9-8"><a href="#cb9-8" tabindex="-1"></a>PRECASTObj <span class="ot">&lt;-</span> <span class="fu">AddAdjList</span>(PRECASTObj, <span class="at">platform =</span> <span class="st">&quot;Visium&quot;</span>)</span>
<span id="cb9-9"><a href="#cb9-9" tabindex="-1"></a></span>
<span id="cb9-10"><a href="#cb9-10" tabindex="-1"></a><span class="do">## Add a model setting in advance for a PRECASTObj object. verbose =TRUE helps outputing the</span></span>
<span id="cb9-11"><a href="#cb9-11" tabindex="-1"></a><span class="do">## information in the algorithm.</span></span>
<span id="cb9-12"><a href="#cb9-12" tabindex="-1"></a>PRECASTObj <span class="ot">&lt;-</span> <span class="fu">AddParSetting</span>(PRECASTObj, <span class="at">Sigma_equal =</span> <span class="cn">FALSE</span>, <span class="at">verbose =</span> <span class="cn">TRUE</span>, <span class="at">maxIter =</span> <span class="dv">30</span>)</span></code></pre></div>
</div>
</div>
<div id="fit-precast-using-this-data" class="section level2">
<h2>Fit PRECAST using this data</h2>
<div id="fit-precast" class="section level3">
<h3>Fit PRECAST</h3>
<p>For function <code>PRECAST</code>, users can specify the number of
clusters <span class="math inline">\(K\)</span> or set <code>K</code> to
be an integer vector by using modified BIC(MBIC) to determine <span class="math inline">\(K\)</span>. First, we try using user-specified
number of clusters. For convenience, we give the selected number of
clusters by MBIC (K=14).</p>
<div class="sourceCode" id="cb10"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" tabindex="-1"></a><span class="do">### Given K</span></span>
<span id="cb10-2"><a href="#cb10-2" tabindex="-1"></a>PRECASTObj <span class="ot">&lt;-</span> <span class="fu">PRECAST</span>(PRECASTObj, <span class="at">K =</span> <span class="dv">14</span>)</span></code></pre></div>
<p>Select a best model if <span class="math inline">\(K\)</span> is an
integer vector. Even if <span class="math inline">\(K\)</span> is a
scalar, this step is also neccessary to re-organize the results in
<code>PRECASTObj</code> object.</p>
<div class="sourceCode" id="cb11"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1" tabindex="-1"></a><span class="do">## backup the fitting results in resList</span></span>
<span id="cb11-2"><a href="#cb11-2" tabindex="-1"></a>resList <span class="ot">&lt;-</span> PRECASTObj<span class="sc">@</span>resList</span>
<span id="cb11-3"><a href="#cb11-3" tabindex="-1"></a>PRECASTObj <span class="ot">&lt;-</span> <span class="fu">SelectModel</span>(PRECASTObj)</span></code></pre></div>
<p>Integrate the two samples using the <code>IntegrateSpaData</code>
function. For computational efficiency, this function exclusively
integrates the variable genes. Specifically, in cases where users do not
specify the <code>PRECASTObj@seuList</code> or <code>seuList</code>
argument within the <code>IntegrateSpaData</code> function, it
automatically focuses on integrating only the variable genes. The
default setting for <code>PRECASTObj@seuList</code> is <code>NULL</code>
when <code>rawData.preserve</code> in <code>CreatePRECASTObject</code>
is set to <code>FALSE</code>. For instance:</p>
<div class="sourceCode" id="cb12"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1" tabindex="-1"></a><span class="fu">print</span>(PRECASTObj<span class="sc">@</span>seuList)</span>
<span id="cb12-2"><a href="#cb12-2" tabindex="-1"></a>seuInt <span class="ot">&lt;-</span> <span class="fu">IntegrateSpaData</span>(PRECASTObj, <span class="at">species =</span> <span class="st">&quot;Human&quot;</span>)</span>
<span id="cb12-3"><a href="#cb12-3" tabindex="-1"></a>seuInt</span>
<span id="cb12-4"><a href="#cb12-4" tabindex="-1"></a><span class="do">## The low-dimensional embeddings obtained by PRECAST are saved in PRECAST reduction slot.</span></span></code></pre></div>
<details>
<summary>
<strong>Integrating all genes</strong>
</summary>
<p>There are two ways to use <code>IntegrateSpaData</code> integrating
all genes, which will require more memory. We recommand running for all
genes on server. The first one is to set value for
<code>PRECASTObj@seuList</code>.</p>
<div class="sourceCode" id="cb13"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1" tabindex="-1"></a><span class="do">## assign the raw Seurat list object to it For illustration, we generate a new seuList with</span></span>
<span id="cb13-2"><a href="#cb13-2" tabindex="-1"></a><span class="do">## more genes; For integrating all genes, users can set `seuList &lt;- bc2`.</span></span>
<span id="cb13-3"><a href="#cb13-3" tabindex="-1"></a>genes <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="fu">row.names</span>(PRECASTObj<span class="sc">@</span>seulist[[<span class="dv">1</span>]]), <span class="fu">row.names</span>(bc2[[<span class="dv">1</span>]])[<span class="dv">1</span><span class="sc">:</span><span class="dv">10</span>])</span>
<span id="cb13-4"><a href="#cb13-4" tabindex="-1"></a>seuList <span class="ot">&lt;-</span> <span class="fu">lapply</span>(bc2, <span class="cf">function</span>(x) x[genes, ])</span>
<span id="cb13-5"><a href="#cb13-5" tabindex="-1"></a>PRECASTObj<span class="sc">@</span>seuList <span class="ot">&lt;-</span> seuList  <span class="co"># </span></span>
<span id="cb13-6"><a href="#cb13-6" tabindex="-1"></a>seuInt <span class="ot">&lt;-</span> <span class="fu">IntegrateSpaData</span>(PRECASTObj, <span class="at">species =</span> <span class="st">&quot;Human&quot;</span>)</span>
<span id="cb13-7"><a href="#cb13-7" tabindex="-1"></a>seuInt</span></code></pre></div>
<p>The second method is to set a value for the argument
<code>seuList</code>:</p>
<div class="sourceCode" id="cb14"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1" tabindex="-1"></a>PRECASTObj<span class="sc">@</span>seuList <span class="ot">&lt;-</span> <span class="cn">NULL</span></span>
<span id="cb14-2"><a href="#cb14-2" tabindex="-1"></a><span class="do">## At the same time, we can set subsampling to speed up the computation.</span></span>
<span id="cb14-3"><a href="#cb14-3" tabindex="-1"></a>seuInt <span class="ot">&lt;-</span> <span class="fu">IntegrateSpaData</span>(PRECASTObj, <span class="at">species =</span> <span class="st">&quot;Human&quot;</span>, <span class="at">seuList =</span> seuList, <span class="at">subsample_rate =</span> <span class="fl">0.5</span>)</span>
<span id="cb14-4"><a href="#cb14-4" tabindex="-1"></a>seuInt</span></code></pre></div>
</details>
<p>First, user can choose a beautiful color schema using
<code>chooseColors()</code>.</p>
<div class="sourceCode" id="cb15"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1" tabindex="-1"></a>cols_cluster <span class="ot">&lt;-</span> <span class="fu">chooseColors</span>(<span class="at">palettes_name =</span> <span class="st">&quot;Classic 20&quot;</span>, <span class="at">n_colors =</span> <span class="dv">14</span>, <span class="at">plot_colors =</span> <span class="cn">FALSE</span>)</span></code></pre></div>
<p>Show the spatial scatter plot for clusters</p>
<div class="sourceCode" id="cb16"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb16-1"><a href="#cb16-1" tabindex="-1"></a></span>
<span id="cb16-2"><a href="#cb16-2" tabindex="-1"></a>p12 <span class="ot">&lt;-</span> <span class="fu">SpaPlot</span>(seuInt, <span class="at">item =</span> <span class="st">&quot;cluster&quot;</span>, <span class="at">batch =</span> <span class="cn">NULL</span>, <span class="at">point_size =</span> <span class="dv">1</span>, <span class="at">cols =</span> cols_cluster, <span class="at">combine =</span> <span class="cn">TRUE</span>,</span>
<span id="cb16-3"><a href="#cb16-3" tabindex="-1"></a>    <span class="at">nrow.legend =</span> <span class="dv">7</span>)</span>
<span id="cb16-4"><a href="#cb16-4" tabindex="-1"></a>p12</span>
<span id="cb16-5"><a href="#cb16-5" tabindex="-1"></a><span class="co"># users can plot each sample by setting combine=FALSE</span></span></code></pre></div>
<p>Users can re-plot the above figures for specific need by returning a
ggplot list object. For example, we plot the spatial heatmap using a
common legend.</p>
<div class="sourceCode" id="cb17"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb17-1"><a href="#cb17-1" tabindex="-1"></a>pList <span class="ot">&lt;-</span> <span class="fu">SpaPlot</span>(seuInt, <span class="at">item =</span> <span class="st">&quot;cluster&quot;</span>, <span class="at">batch =</span> <span class="cn">NULL</span>, <span class="at">point_size =</span> <span class="dv">1</span>, <span class="at">cols =</span> cols_cluster, <span class="at">combine =</span> <span class="cn">FALSE</span>,</span>
<span id="cb17-2"><a href="#cb17-2" tabindex="-1"></a>    <span class="at">nrow.legend =</span> <span class="dv">7</span>)</span>
<span id="cb17-3"><a href="#cb17-3" tabindex="-1"></a><span class="fu">drawFigs</span>(pList, <span class="at">layout.dim =</span> <span class="fu">c</span>(<span class="dv">1</span>, <span class="dv">2</span>), <span class="at">common.legend =</span> <span class="cn">TRUE</span>, <span class="at">legend.position =</span> <span class="st">&quot;right&quot;</span>, <span class="at">align =</span> <span class="st">&quot;hv&quot;</span>)</span></code></pre></div>
<p>Show the spatial UMAP/tNSE RGB plot to illustrate the performance in
extracting features.</p>
<div class="sourceCode" id="cb18"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb18-1"><a href="#cb18-1" tabindex="-1"></a>seuInt <span class="ot">&lt;-</span> <span class="fu">AddUMAP</span>(seuInt)</span>
<span id="cb18-2"><a href="#cb18-2" tabindex="-1"></a>p13 <span class="ot">&lt;-</span> <span class="fu">SpaPlot</span>(seuInt, <span class="at">batch =</span> <span class="cn">NULL</span>, <span class="at">item =</span> <span class="st">&quot;RGB_UMAP&quot;</span>, <span class="at">point_size =</span> <span class="dv">2</span>, <span class="at">combine =</span> <span class="cn">TRUE</span>, <span class="at">text_size =</span> <span class="dv">15</span>)</span>
<span id="cb18-3"><a href="#cb18-3" tabindex="-1"></a>p13</span>
<span id="cb18-4"><a href="#cb18-4" tabindex="-1"></a><span class="co"># seuInt &lt;- AddTSNE(seuInt) SpaPlot(seuInt, batch=NULL,item=&#39;RGB_TSNE&#39;,point_size=2,</span></span>
<span id="cb18-5"><a href="#cb18-5" tabindex="-1"></a><span class="co"># combine=T, text_size=15)</span></span></code></pre></div>
<p>Show the tSNE plot based on the extracted features from PRECAST to
check the performance of integration.</p>
<div class="sourceCode" id="cb19"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb19-1"><a href="#cb19-1" tabindex="-1"></a>seuInt <span class="ot">&lt;-</span> <span class="fu">AddTSNE</span>(seuInt, <span class="at">n_comp =</span> <span class="dv">2</span>)</span>
<span id="cb19-2"><a href="#cb19-2" tabindex="-1"></a>p1 <span class="ot">&lt;-</span> <span class="fu">dimPlot</span>(seuInt, <span class="at">item =</span> <span class="st">&quot;cluster&quot;</span>, <span class="at">point_size =</span> <span class="fl">0.5</span>, <span class="at">font_family =</span> <span class="st">&quot;serif&quot;</span>, <span class="at">cols =</span> cols_cluster,</span>
<span id="cb19-3"><a href="#cb19-3" tabindex="-1"></a>    <span class="at">border_col =</span> <span class="st">&quot;gray10&quot;</span>, <span class="at">nrow.legend =</span> <span class="dv">14</span>, <span class="at">legend_pos =</span> <span class="st">&quot;right&quot;</span>)  <span class="co"># Times New Roman</span></span>
<span id="cb19-4"><a href="#cb19-4" tabindex="-1"></a>p2 <span class="ot">&lt;-</span> <span class="fu">dimPlot</span>(seuInt, <span class="at">item =</span> <span class="st">&quot;batch&quot;</span>, <span class="at">point_size =</span> <span class="fl">0.5</span>, <span class="at">font_family =</span> <span class="st">&quot;serif&quot;</span>, <span class="at">legend_pos =</span> <span class="st">&quot;right&quot;</span>)</span>
<span id="cb19-5"><a href="#cb19-5" tabindex="-1"></a></span>
<span id="cb19-6"><a href="#cb19-6" tabindex="-1"></a><span class="fu">drawFigs</span>(<span class="fu">list</span>(p1, p2), <span class="at">layout.dim =</span> <span class="fu">c</span>(<span class="dv">1</span>, <span class="dv">2</span>), <span class="at">legend.position =</span> <span class="st">&quot;right&quot;</span>, <span class="at">align =</span> <span class="st">&quot;hv&quot;</span>)</span></code></pre></div>
<p>Combined differential expression analysis</p>
<div class="sourceCode" id="cb20"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb20-1"><a href="#cb20-1" tabindex="-1"></a><span class="fu">library</span>(Seurat)</span>
<span id="cb20-2"><a href="#cb20-2" tabindex="-1"></a>dat_deg <span class="ot">&lt;-</span> <span class="fu">FindAllMarkers</span>(seuInt)</span>
<span id="cb20-3"><a href="#cb20-3" tabindex="-1"></a><span class="fu">library</span>(dplyr)</span>
<span id="cb20-4"><a href="#cb20-4" tabindex="-1"></a>n <span class="ot">&lt;-</span> <span class="dv">10</span></span>
<span id="cb20-5"><a href="#cb20-5" tabindex="-1"></a>dat_deg <span class="sc">%&gt;%</span></span>
<span id="cb20-6"><a href="#cb20-6" tabindex="-1"></a>    <span class="fu">group_by</span>(cluster) <span class="sc">%&gt;%</span></span>
<span id="cb20-7"><a href="#cb20-7" tabindex="-1"></a>    <span class="fu">top_n</span>(<span class="at">n =</span> n, <span class="at">wt =</span> avg_log2FC) <span class="ot">-&gt;</span> top10</span>
<span id="cb20-8"><a href="#cb20-8" tabindex="-1"></a></span>
<span id="cb20-9"><a href="#cb20-9" tabindex="-1"></a>seuInt <span class="ot">&lt;-</span> <span class="fu">ScaleData</span>(seuInt)</span>
<span id="cb20-10"><a href="#cb20-10" tabindex="-1"></a>seus <span class="ot">&lt;-</span> <span class="fu">subset</span>(seuInt, <span class="at">downsample =</span> <span class="dv">400</span>)</span></code></pre></div>
<p>Plot DE genesâ heatmap for each spatial domain identified by
PRECAST.</p>
<div class="sourceCode" id="cb21"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb21-1"><a href="#cb21-1" tabindex="-1"></a>color_id <span class="ot">&lt;-</span> <span class="fu">as.numeric</span>(<span class="fu">levels</span>(<span class="fu">Idents</span>(seus)))</span>
<span id="cb21-2"><a href="#cb21-2" tabindex="-1"></a><span class="fu">library</span>(ggplot2)</span>
<span id="cb21-3"><a href="#cb21-3" tabindex="-1"></a><span class="do">## HeatMap</span></span>
<span id="cb21-4"><a href="#cb21-4" tabindex="-1"></a>p1 <span class="ot">&lt;-</span> <span class="fu">doHeatmap</span>(seus, <span class="at">features =</span> top10<span class="sc">$</span>gene, <span class="at">cell_label =</span> <span class="st">&quot;Domain&quot;</span>, <span class="at">grp_label =</span> F, <span class="at">grp_color =</span> cols_cluster[color_id],</span>
<span id="cb21-5"><a href="#cb21-5" tabindex="-1"></a>    <span class="at">pt_size =</span> <span class="dv">6</span>, <span class="at">slot =</span> <span class="st">&quot;scale.data&quot;</span>) <span class="sc">+</span> <span class="fu">theme</span>(<span class="at">legend.text =</span> <span class="fu">element_text</span>(<span class="at">size =</span> <span class="dv">10</span>), <span class="at">legend.title =</span> <span class="fu">element_text</span>(<span class="at">size =</span> <span class="dv">13</span>,</span>
<span id="cb21-6"><a href="#cb21-6" tabindex="-1"></a>    <span class="at">face =</span> <span class="st">&quot;bold&quot;</span>), <span class="at">axis.text.y =</span> <span class="fu">element_text</span>(<span class="at">size =</span> <span class="dv">5</span>, <span class="at">face =</span> <span class="st">&quot;italic&quot;</span>, <span class="at">family =</span> <span class="st">&quot;serif&quot;</span>))</span>
<span id="cb21-7"><a href="#cb21-7" tabindex="-1"></a>p1</span></code></pre></div>
<details>
<summary>
<strong>Session Info</strong>
</summary>
<div class="sourceCode" id="cb22"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb22-1"><a href="#cb22-1" tabindex="-1"></a><span class="fu">sessionInfo</span>()</span>
<span id="cb22-2"><a href="#cb22-2" tabindex="-1"></a><span class="co">#&gt; R version 4.4.1 (2024-06-14 ucrt)</span></span>
<span id="cb22-3"><a href="#cb22-3" tabindex="-1"></a><span class="co">#&gt; Platform: x86_64-w64-mingw32/x64</span></span>
<span id="cb22-4"><a href="#cb22-4" tabindex="-1"></a><span class="co">#&gt; Running under: Windows 11 x64 (build 26100)</span></span>
<span id="cb22-5"><a href="#cb22-5" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb22-6"><a href="#cb22-6" tabindex="-1"></a><span class="co">#&gt; Matrix products: default</span></span>
<span id="cb22-7"><a href="#cb22-7" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb22-8"><a href="#cb22-8" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb22-9"><a href="#cb22-9" tabindex="-1"></a><span class="co">#&gt; locale:</span></span>
<span id="cb22-10"><a href="#cb22-10" tabindex="-1"></a><span class="co">#&gt; [1] LC_COLLATE=C                               </span></span>
<span id="cb22-11"><a href="#cb22-11" tabindex="-1"></a><span class="co">#&gt; [2] LC_CTYPE=Chinese (Simplified)_China.utf8   </span></span>
<span id="cb22-12"><a href="#cb22-12" tabindex="-1"></a><span class="co">#&gt; [3] LC_MONETARY=Chinese (Simplified)_China.utf8</span></span>
<span id="cb22-13"><a href="#cb22-13" tabindex="-1"></a><span class="co">#&gt; [4] LC_NUMERIC=C                               </span></span>
<span id="cb22-14"><a href="#cb22-14" tabindex="-1"></a><span class="co">#&gt; [5] LC_TIME=Chinese (Simplified)_China.utf8    </span></span>
<span id="cb22-15"><a href="#cb22-15" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb22-16"><a href="#cb22-16" tabindex="-1"></a><span class="co">#&gt; time zone: Asia/Shanghai</span></span>
<span id="cb22-17"><a href="#cb22-17" tabindex="-1"></a><span class="co">#&gt; tzcode source: internal</span></span>
<span id="cb22-18"><a href="#cb22-18" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb22-19"><a href="#cb22-19" tabindex="-1"></a><span class="co">#&gt; attached base packages:</span></span>
<span id="cb22-20"><a href="#cb22-20" tabindex="-1"></a><span class="co">#&gt; [1] stats     graphics  grDevices utils     datasets  methods   base     </span></span>
<span id="cb22-21"><a href="#cb22-21" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb22-22"><a href="#cb22-22" tabindex="-1"></a><span class="co">#&gt; loaded via a namespace (and not attached):</span></span>
<span id="cb22-23"><a href="#cb22-23" tabindex="-1"></a><span class="co">#&gt;  [1] digest_0.6.37     R6_2.5.1          fastmap_1.2.0     xfun_0.47        </span></span>
<span id="cb22-24"><a href="#cb22-24" tabindex="-1"></a><span class="co">#&gt;  [5] cachem_1.1.0      knitr_1.48        htmltools_0.5.8.1 rmarkdown_2.28   </span></span>
<span id="cb22-25"><a href="#cb22-25" tabindex="-1"></a><span class="co">#&gt;  [9] lifecycle_1.0.4   cli_3.6.3         sass_0.4.9        jquerylib_0.1.4  </span></span>
<span id="cb22-26"><a href="#cb22-26" tabindex="-1"></a><span class="co">#&gt; [13] compiler_4.4.1    rstudioapi_0.16.0 tools_4.4.1       evaluate_1.0.0   </span></span>
<span id="cb22-27"><a href="#cb22-27" tabindex="-1"></a><span class="co">#&gt; [17] bslib_0.8.0       yaml_2.3.10       formatR_1.14      rlang_1.1.4      </span></span>
<span id="cb22-28"><a href="#cb22-28" tabindex="-1"></a><span class="co">#&gt; [21] jsonlite_1.8.9</span></span></code></pre></div>
</details>
</div>
</div>



<!-- code folding -->


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
