<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />

<meta name="viewport" content="width=device-width, initial-scale=1" />

<meta name="author" content="Wei Liu" />

<meta name="date" content="2025-10-01" />

<title>PRECAST: simulation</title>

<script>// Pandoc 2.9 adds attributes on both header and div. We remove the former (to
// be compatible with the behavior of Pandoc < 2.8).
document.addEventListener('DOMContentLoaded', function(e) {
  var hs = document.querySelectorAll("div.section[class*='level'] > :first-child");
  var i, h, a;
  for (i = 0; i < hs.length; i++) {
    h = hs[i];
    if (!/^h[1-6]$/i.test(h.tagName)) continue;  // it should be a header h1-h6
    a = h.attributes;
    while (a.length > 0) h.removeAttribute(a[0].name);
  }
});
</script>

<style type="text/css">
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
span.underline{text-decoration: underline;}
div.column{display: inline-block; vertical-align: top; width: 50%;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
</style>



<style type="text/css">
code {
white-space: pre;
}
.sourceCode {
overflow: visible;
}
</style>
<style type="text/css" data-origin="pandoc">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
{ counter-reset: source-line 0; }
pre.numberSource code > span
{ position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
{ content: counter(source-line);
position: relative; left: -1em; text-align: right; vertical-align: baseline;
border: none; display: inline-block;
-webkit-touch-callout: none; -webkit-user-select: none;
-khtml-user-select: none; -moz-user-select: none;
-ms-user-select: none; user-select: none;
padding: 0 4px; width: 4em;
color: #aaaaaa;
}
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa; padding-left: 4px; }
div.sourceCode
{ }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } 
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } 
code span.at { color: #7d9029; } 
code span.bn { color: #40a070; } 
code span.bu { color: #008000; } 
code span.cf { color: #007020; font-weight: bold; } 
code span.ch { color: #4070a0; } 
code span.cn { color: #880000; } 
code span.co { color: #60a0b0; font-style: italic; } 
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } 
code span.do { color: #ba2121; font-style: italic; } 
code span.dt { color: #902000; } 
code span.dv { color: #40a070; } 
code span.er { color: #ff0000; font-weight: bold; } 
code span.ex { } 
code span.fl { color: #40a070; } 
code span.fu { color: #06287e; } 
code span.im { color: #008000; font-weight: bold; } 
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } 
code span.kw { color: #007020; font-weight: bold; } 
code span.op { color: #666666; } 
code span.ot { color: #007020; } 
code span.pp { color: #bc7a00; } 
code span.sc { color: #4070a0; } 
code span.ss { color: #bb6688; } 
code span.st { color: #4070a0; } 
code span.va { color: #19177c; } 
code span.vs { color: #4070a0; } 
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } 
</style>
<script>
// apply pandoc div.sourceCode style to pre.sourceCode instead
(function() {
  var sheets = document.styleSheets;
  for (var i = 0; i < sheets.length; i++) {
    if (sheets[i].ownerNode.dataset["origin"] !== "pandoc") continue;
    try { var rules = sheets[i].cssRules; } catch (e) { continue; }
    var j = 0;
    while (j < rules.length) {
      var rule = rules[j];
      // check if there is a div.sourceCode rule
      if (rule.type !== rule.STYLE_RULE || rule.selectorText !== "div.sourceCode") {
        j++;
        continue;
      }
      var style = rule.style.cssText;
      // check if color or background-color is set
      if (rule.style.color === '' && rule.style.backgroundColor === '') {
        j++;
        continue;
      }
      // replace div.sourceCode by a pre.sourceCode rule
      sheets[i].deleteRule(j);
      sheets[i].insertRule('pre.sourceCode{' + style + '}', j);
    }
  }
})();
</script>




<style type="text/css">body {
background-color: #fff;
margin: 1em auto;
max-width: 700px;
overflow: visible;
padding-left: 2em;
padding-right: 2em;
font-family: "Open Sans", "Helvetica Neue", Helvetica, Arial, sans-serif;
font-size: 14px;
line-height: 1.35;
}
#TOC {
clear: both;
margin: 0 0 10px 10px;
padding: 4px;
width: 400px;
border: 1px solid #CCCCCC;
border-radius: 5px;
background-color: #f6f6f6;
font-size: 13px;
line-height: 1.3;
}
#TOC .toctitle {
font-weight: bold;
font-size: 15px;
margin-left: 5px;
}
#TOC ul {
padding-left: 40px;
margin-left: -1.5em;
margin-top: 5px;
margin-bottom: 5px;
}
#TOC ul ul {
margin-left: -2em;
}
#TOC li {
line-height: 16px;
}
table {
margin: 1em auto;
border-width: 1px;
border-color: #DDDDDD;
border-style: outset;
border-collapse: collapse;
}
table th {
border-width: 2px;
padding: 5px;
border-style: inset;
}
table td {
border-width: 1px;
border-style: inset;
line-height: 18px;
padding: 5px 5px;
}
table, table th, table td {
border-left-style: none;
border-right-style: none;
}
table thead, table tr.even {
background-color: #f7f7f7;
}
p {
margin: 0.5em 0;
}
blockquote {
background-color: #f6f6f6;
padding: 0.25em 0.75em;
}
hr {
border-style: solid;
border: none;
border-top: 1px solid #777;
margin: 28px 0;
}
dl {
margin-left: 0;
}
dl dd {
margin-bottom: 13px;
margin-left: 13px;
}
dl dt {
font-weight: bold;
}
ul {
margin-top: 0;
}
ul li {
list-style: circle outside;
}
ul ul {
margin-bottom: 0;
}
pre, code {
background-color: #f7f7f7;
border-radius: 3px;
color: #333;
white-space: pre-wrap; 
}
pre {
border-radius: 3px;
margin: 5px 0px 10px 0px;
padding: 10px;
}
pre:not([class]) {
background-color: #f7f7f7;
}
code {
font-family: Consolas, Monaco, 'Courier New', monospace;
font-size: 85%;
}
p > code, li > code {
padding: 2px 0px;
}
div.figure {
text-align: center;
}
img {
background-color: #FFFFFF;
padding: 2px;
border: 1px solid #DDDDDD;
border-radius: 3px;
border: 1px solid #CCCCCC;
margin: 0 5px;
}
h1 {
margin-top: 0;
font-size: 35px;
line-height: 40px;
}
h2 {
border-bottom: 4px solid #f7f7f7;
padding-top: 10px;
padding-bottom: 2px;
font-size: 145%;
}
h3 {
border-bottom: 2px solid #f7f7f7;
padding-top: 10px;
font-size: 120%;
}
h4 {
border-bottom: 1px solid #f7f7f7;
margin-left: 8px;
font-size: 105%;
}
h5, h6 {
border-bottom: 1px solid #ccc;
font-size: 105%;
}
a {
color: #0033dd;
text-decoration: none;
}
a:hover {
color: #6666ff; }
a:visited {
color: #800080; }
a:visited:hover {
color: #BB00BB; }
a[href^="http:"] {
text-decoration: underline; }
a[href^="https:"] {
text-decoration: underline; }

code > span.kw { color: #555; font-weight: bold; } 
code > span.dt { color: #902000; } 
code > span.dv { color: #40a070; } 
code > span.bn { color: #d14; } 
code > span.fl { color: #d14; } 
code > span.ch { color: #d14; } 
code > span.st { color: #d14; } 
code > span.co { color: #888888; font-style: italic; } 
code > span.ot { color: #007020; } 
code > span.al { color: #ff0000; font-weight: bold; } 
code > span.fu { color: #900; font-weight: bold; } 
code > span.er { color: #a61717; background-color: #e3d2d2; } 
</style>




</head>

<body>




<h1 class="title toc-ignore">PRECAST: simulation</h1>
<h4 class="author">Wei Liu</h4>
<h4 class="date">2025-10-01</h4>



<p>This vignette introduces the PRECAST workflow for the analysis of
integrating multiple spatial transcriptomics dataset. The workflow
consists of three steps</p>
<ul>
<li>Independent preprocessing and model setting</li>
<li>Probabilistic embedding, clustering and alignment using PRECAST
model</li>
<li>Downstream analysis (i.e.Â visualization of clusters and embeddings,
combined differential expression analysis)</li>
</ul>
<p>We demonstrate the use of PRECAST to three simulated Visium data that
are <a href="https://github.com/feiyoung/PRECAST/tree/main/vignettes_data">here</a>,
which can be downloaded to the current working path by the following
command:</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" tabindex="-1"></a>githubURL <span class="ot">&lt;-</span> <span class="st">&quot;https://github.com/feiyoung/PRECAST/blob/main/vignettes_data/data_simu.rda?raw=true&quot;</span></span>
<span id="cb1-2"><a href="#cb1-2" tabindex="-1"></a><span class="fu">download.file</span>(githubURL,<span class="st">&quot;data_simu.rda&quot;</span>,<span class="at">mode=</span><span class="st">&#39;wb&#39;</span>)</span></code></pre></div>
<p>Then load to R</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" tabindex="-1"></a><span class="fu">load</span>(<span class="st">&quot;data_simu.rda&quot;</span>)</span></code></pre></div>
<p>The package can be loaded with the command:</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" tabindex="-1"></a><span class="fu">library</span>(PRECAST)</span>
<span id="cb3-2"><a href="#cb3-2" tabindex="-1"></a><span class="fu">library</span>(Seurat)</span></code></pre></div>
<div id="load-the-simulated-data" class="section level2">
<h2>Load the simulated data</h2>
<p>First, we view the the three simulated spatial transcriptomics data
with Visium platform.</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" tabindex="-1"></a>data_simu <span class="do">## a list including three Seurat object with default assay: RNA</span></span></code></pre></div>
<p>Check the content in <code>data_simu</code>.</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" tabindex="-1"></a><span class="fu">head</span>(data_simu[[<span class="dv">1</span>]])</span></code></pre></div>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" tabindex="-1"></a><span class="fu">row.names</span>(data_simu[[<span class="dv">1</span>]])[<span class="dv">1</span><span class="sc">:</span><span class="dv">10</span>]</span></code></pre></div>
</div>
<div id="create-a-precastobject-object" class="section level2">
<h2>Create a PRECASTObject object</h2>
<p>We show how to create a PRECASTObject object step by step. First, we
create a Seurat list object using the count matrix and meta data of each
data batch. Although <code>data_simu</code> is a prepared Seurat list
object, we re-create a same objcet seuList to show the details.</p>
<ul>
<li>Note: the spatial coordinates must be contained in the meta data and
named as <code>row</code> and <code>col</code>, which benefits the
identification of spaital coordinates by PRECAST.</li>
</ul>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" tabindex="-1"></a><span class="do">## Get the gene-by-spot read count matrices</span></span>
<span id="cb7-2"><a href="#cb7-2" tabindex="-1"></a>countList <span class="ot">&lt;-</span> <span class="fu">lapply</span>(data_simu, <span class="cf">function</span>(x){</span>
<span id="cb7-3"><a href="#cb7-3" tabindex="-1"></a>  assay <span class="ot">&lt;-</span> <span class="fu">DefaultAssay</span>(x)</span>
<span id="cb7-4"><a href="#cb7-4" tabindex="-1"></a>  <span class="fu">GetAssayData</span>(x, <span class="at">assay =</span> assay, <span class="at">slot=</span><span class="st">&#39;counts&#39;</span>)</span>
<span id="cb7-5"><a href="#cb7-5" tabindex="-1"></a>  </span>
<span id="cb7-6"><a href="#cb7-6" tabindex="-1"></a>} )</span>
<span id="cb7-7"><a href="#cb7-7" tabindex="-1"></a></span>
<span id="cb7-8"><a href="#cb7-8" tabindex="-1"></a><span class="do">## Check the spatial coordinates: Yes, they are named as &quot;row&quot; and &quot;col&quot;!</span></span>
<span id="cb7-9"><a href="#cb7-9" tabindex="-1"></a><span class="fu">head</span>(data_simu[[<span class="dv">1</span>]]<span class="sc">@</span>meta.data)</span>
<span id="cb7-10"><a href="#cb7-10" tabindex="-1"></a></span>
<span id="cb7-11"><a href="#cb7-11" tabindex="-1"></a><span class="do">## Get the meta data of each spot for each data batch</span></span>
<span id="cb7-12"><a href="#cb7-12" tabindex="-1"></a>metadataList <span class="ot">&lt;-</span> <span class="fu">lapply</span>(data_simu, <span class="cf">function</span>(x) x<span class="sc">@</span>meta.data)</span>
<span id="cb7-13"><a href="#cb7-13" tabindex="-1"></a></span>
<span id="cb7-14"><a href="#cb7-14" tabindex="-1"></a></span>
<span id="cb7-15"><a href="#cb7-15" tabindex="-1"></a><span class="do">## ensure the row.names of metadata in metaList are the same as that of colnames count matrix in countList</span></span>
<span id="cb7-16"><a href="#cb7-16" tabindex="-1"></a>M <span class="ot">&lt;-</span> <span class="fu">length</span>(countList)</span>
<span id="cb7-17"><a href="#cb7-17" tabindex="-1"></a><span class="cf">for</span>(r <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span>M){</span>
<span id="cb7-18"><a href="#cb7-18" tabindex="-1"></a>  <span class="fu">row.names</span>(metadataList[[r]]) <span class="ot">&lt;-</span> <span class="fu">colnames</span>(countList[[r]])</span>
<span id="cb7-19"><a href="#cb7-19" tabindex="-1"></a>}</span>
<span id="cb7-20"><a href="#cb7-20" tabindex="-1"></a></span>
<span id="cb7-21"><a href="#cb7-21" tabindex="-1"></a></span>
<span id="cb7-22"><a href="#cb7-22" tabindex="-1"></a><span class="do">## Create the Seurat list  object</span></span>
<span id="cb7-23"><a href="#cb7-23" tabindex="-1"></a></span>
<span id="cb7-24"><a href="#cb7-24" tabindex="-1"></a>seuList <span class="ot">&lt;-</span> <span class="fu">list</span>()</span>
<span id="cb7-25"><a href="#cb7-25" tabindex="-1"></a><span class="cf">for</span>(r <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span>M){</span>
<span id="cb7-26"><a href="#cb7-26" tabindex="-1"></a>  seuList[[r]] <span class="ot">&lt;-</span> <span class="fu">CreateSeuratObject</span>(<span class="at">counts =</span> countList[[r]], <span class="at">meta.data=</span>metadataList[[r]], <span class="at">project =</span> <span class="st">&quot;PRECASTsimu&quot;</span>)</span>
<span id="cb7-27"><a href="#cb7-27" tabindex="-1"></a>}</span></code></pre></div>
<div id="prepare-the-precastobject-with-preprocessing-step." class="section level3">
<h3>Prepare the PRECASTObject with preprocessing step.</h3>
<p>Next, we use <code>CreatePRECASTObject()</code> to create a
PRECASTObject based on the Seurat list object <code>seuList</code>. This
function will do three things:</p>
<ul>
<li><ol style="list-style-type: decimal">
<li>Filter low-quality spots and genes, controlled by the arguments
<code>premin.features</code> and <code>premin.spots</code>,
respectively; the spots are retained in raw data (seuList) with at least
premin.features number of nonzero-count features (genes), and the genes
are retained in raw data (seuList) with at least
<code>premin.spots</code> number of spots. To ease presentation, we
denote the filtered Seurat list object as data_filter1.</li>
</ol></li>
<li><ol start="2" style="list-style-type: decimal">
<li>Select the top 2,000 variable genes (by setting
<code>gene.number=2000</code>) for each data batch using
<code>FindSVGs()</code> function in <code>DR.SC</code> package for
spatially variable genes or <code>FindVariableFeatures()</code> function
in <code>Seurat</code> package for highly variable genes. Next, we
prioritized genes based on the number of times they were selected as
variable genes in all samples and chose the top 2,000 genes. Then denote
the Seurat list object as data_filter2, where only 2,000 genes are
retained.</li>
</ol></li>
<li><ol start="3" style="list-style-type: decimal">
<li>Conduct strict quality control for data_filter2 by filtering spots
and genes, controlled by the arguments <code>postmin.features</code> and
<code>postmin.spots</code>, respectively; the spots are retained with at
least <code>post.features</code> nonzero counts across genes; the
features (genes) are retained with at least <code>postmin.spots</code>
number of nonzero-count spots. Usually, no genes are filltered because
these genes are variable genes.</li>
</ol></li>
</ul>
<p>If the argument <code>customGenelist</code> is not <code>NULL</code>,
then this function only does (3) based on <code>customGenelist</code>
gene list.</p>
<p>In this simulated dataset, we donât require to select genes, thus, we
set <code>customGenelist=row.names(seuList[[1]])</code>, representing
the user-defined gene list. User can retain the raw seurat list object
by setting <code>rawData.preserve = TRUE</code>.</p>
<div class="sourceCode" id="cb8"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" tabindex="-1"></a></span>
<span id="cb8-2"><a href="#cb8-2" tabindex="-1"></a><span class="do">## Create PRECASTObject</span></span>
<span id="cb8-3"><a href="#cb8-3" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">2022</span>)</span>
<span id="cb8-4"><a href="#cb8-4" tabindex="-1"></a>PRECASTObj <span class="ot">&lt;-</span>  <span class="fu">CreatePRECASTObject</span>(seuList, <span class="at">customGenelist=</span><span class="fu">row.names</span>(seuList[[<span class="dv">1</span>]]))</span>
<span id="cb8-5"><a href="#cb8-5" tabindex="-1"></a></span>
<span id="cb8-6"><a href="#cb8-6" tabindex="-1"></a><span class="do">## User can retain the raw seuList by the following commond.</span></span>
<span id="cb8-7"><a href="#cb8-7" tabindex="-1"></a><span class="do">##  PRECASTObj &lt;-  CreatePRECASTObject(seuList, customGenelist=row.names(seuList[[1]]), rawData.preserve = TRUE)</span></span></code></pre></div>
</div>
</div>
<div id="fit-precast-using-simulated-data" class="section level2">
<h2>Fit PRECAST using simulated data</h2>
<div id="add-the-model-setting" class="section level3">
<h3>Add the model setting</h3>
<p>Add adjacency matrix list and parameter setting of PRECAST. More
model setting parameters can be found in <code>model_set()</code>.</p>
<div class="sourceCode" id="cb9"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" tabindex="-1"></a><span class="do">## check the number of genes/features after filtering step</span></span>
<span id="cb9-2"><a href="#cb9-2" tabindex="-1"></a>PRECASTObj<span class="sc">@</span>seulist</span>
<span id="cb9-3"><a href="#cb9-3" tabindex="-1"></a></span>
<span id="cb9-4"><a href="#cb9-4" tabindex="-1"></a><span class="do">## seuList is null since the default value `rawData.preserve` is FALSE.</span></span>
<span id="cb9-5"><a href="#cb9-5" tabindex="-1"></a>PRECASTObj<span class="sc">@</span>seuList</span>
<span id="cb9-6"><a href="#cb9-6" tabindex="-1"></a></span>
<span id="cb9-7"><a href="#cb9-7" tabindex="-1"></a><span class="do">## Add adjacency matrix list for a PRECASTObj object to prepare for PRECAST model fitting.</span></span>
<span id="cb9-8"><a href="#cb9-8" tabindex="-1"></a>PRECASTObj <span class="ot">&lt;-</span>  <span class="fu">AddAdjList</span>(PRECASTObj, <span class="at">platform =</span> <span class="st">&quot;Visium&quot;</span>)</span>
<span id="cb9-9"><a href="#cb9-9" tabindex="-1"></a></span>
<span id="cb9-10"><a href="#cb9-10" tabindex="-1"></a><span class="do">## Add a model setting in advance for a PRECASTObj object: verbose =TRUE helps outputing the information in the algorithm; coreNum set the how many cores are used in PRECAST. If you run PRECAST for multiple number of clusters, you can set multiple cores; otherwise, set it to 1. </span></span>
<span id="cb9-11"><a href="#cb9-11" tabindex="-1"></a>PRECASTObj <span class="ot">&lt;-</span> <span class="fu">AddParSetting</span>(PRECASTObj, <span class="at">Sigma_equal=</span><span class="cn">FALSE</span>, <span class="at">maxIter=</span><span class="dv">30</span>, <span class="at">verbose=</span><span class="cn">TRUE</span>, <span class="at">int.model=</span><span class="st">&#39;kmeans&#39;</span>,</span>
<span id="cb9-12"><a href="#cb9-12" tabindex="-1"></a>                             <span class="at">coreNum =</span><span class="dv">1</span>)</span></code></pre></div>
</div>
<div id="fit-precast" class="section level3">
<h3>Fit PRECAST</h3>
<p>For function <code>PRECAST</code>, users can specify the number of
clusters <span class="math inline">\(K\)</span> or set <code>K</code> to
be an integer vector by using modified BIC(MBIC) to determine <span class="math inline">\(K\)</span>. For convenience, we give a single K
here.</p>
<div class="sourceCode" id="cb10"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" tabindex="-1"></a><span class="do">### Given K</span></span>
<span id="cb10-2"><a href="#cb10-2" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">2022</span>)</span>
<span id="cb10-3"><a href="#cb10-3" tabindex="-1"></a>PRECASTObj <span class="ot">&lt;-</span> <span class="fu">PRECAST</span>(PRECASTObj, <span class="at">K=</span><span class="dv">7</span>)</span></code></pre></div>
<details>
<summary>
<strong>Other options</strong>
</summary>
<p>Run for multiple K. Here, we set <code>K=6:9</code>.</p>
<div class="sourceCode" id="cb11"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1" tabindex="-1"></a><span class="do">## Reset  parameters by increasing cores.</span></span>
<span id="cb11-2"><a href="#cb11-2" tabindex="-1"></a>PRECASTObj2 <span class="ot">&lt;-</span> <span class="fu">AddParSetting</span>(PRECASTObj, <span class="at">Sigma_equal=</span><span class="cn">FALSE</span>, <span class="at">maxIter=</span><span class="dv">30</span>, <span class="at">verbose=</span><span class="cn">TRUE</span>,</span>
<span id="cb11-3"><a href="#cb11-3" tabindex="-1"></a>                             <span class="at">coreNum =</span><span class="dv">2</span>)</span>
<span id="cb11-4"><a href="#cb11-4" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">2023</span>)</span>
<span id="cb11-5"><a href="#cb11-5" tabindex="-1"></a>PRECASTObj2 <span class="ot">&lt;-</span> <span class="fu">PRECAST</span>(PRECASTObj2, <span class="at">K=</span><span class="dv">6</span><span class="sc">:</span><span class="dv">7</span>)</span>
<span id="cb11-6"><a href="#cb11-6" tabindex="-1"></a></span>
<span id="cb11-7"><a href="#cb11-7" tabindex="-1"></a>resList2 <span class="ot">&lt;-</span> PRECASTObj2<span class="sc">@</span>resList</span>
<span id="cb11-8"><a href="#cb11-8" tabindex="-1"></a>PRECASTObj2 <span class="ot">&lt;-</span> <span class="fu">SelectModel</span>(PRECASTObj2)</span></code></pre></div>
<ul>
<li>Note: For parallel compuation based on Rcpp on Linux, users require
to use the following system command to set the C_stack unlimited in case
of R Error: C stack usage is too close to the limit.</li>
</ul>
<p><code>ulimit -s unlimited</code></p>
<p>Besides, user can also use different initialization method by setting
<code>int.model</code>, for example, set <code>int.model=NULL</code>;
see the functions <code>AddParSetting()</code> and
<code>model_set()</code> for more details.</p>
</details>
<p>Select a best model and re-organize the results by useing
<code>SelectModel()</code>. Even though <code>K</code> is not a vector,
it is also necessary to run <code>SelectModel()</code> to re-organize
the results in <code>PRECASTObj</code>. The selected best K is 7 by
using command <code>str(PRECASTObj@resList)</code>.</p>
<div class="sourceCode" id="cb12"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1" tabindex="-1"></a><span class="do">## check the fitted results: there are four list for the fitted results of each K (6:9).</span></span>
<span id="cb12-2"><a href="#cb12-2" tabindex="-1"></a><span class="fu">str</span>(PRECASTObj<span class="sc">@</span>resList)</span>
<span id="cb12-3"><a href="#cb12-3" tabindex="-1"></a><span class="do">## backup the fitted results in resList</span></span>
<span id="cb12-4"><a href="#cb12-4" tabindex="-1"></a>resList <span class="ot">&lt;-</span> PRECASTObj<span class="sc">@</span>resList</span>
<span id="cb12-5"><a href="#cb12-5" tabindex="-1"></a><span class="co"># PRECASTObj@resList &lt;- resList</span></span>
<span id="cb12-6"><a href="#cb12-6" tabindex="-1"></a>PRECASTObj <span class="ot">&lt;-</span> <span class="fu">SelectModel</span>(PRECASTObj)</span>
<span id="cb12-7"><a href="#cb12-7" tabindex="-1"></a><span class="do">## check the best and re-organized results</span></span>
<span id="cb12-8"><a href="#cb12-8" tabindex="-1"></a><span class="fu">str</span>(PRECASTObj<span class="sc">@</span>resList) <span class="do">## The selected best K is 7</span></span></code></pre></div>
<p>Use ARI to check the performance of clustering:</p>
<div class="sourceCode" id="cb13"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1" tabindex="-1"></a>true_cluster <span class="ot">&lt;-</span> <span class="fu">lapply</span>(PRECASTObj<span class="sc">@</span>seulist, <span class="cf">function</span>(x) x<span class="sc">$</span>true_cluster)</span>
<span id="cb13-2"><a href="#cb13-2" tabindex="-1"></a><span class="fu">str</span>(true_cluster)</span>
<span id="cb13-3"><a href="#cb13-3" tabindex="-1"></a>mclust<span class="sc">::</span><span class="fu">adjustedRandIndex</span>(<span class="fu">unlist</span>(PRECASTObj<span class="sc">@</span>resList<span class="sc">$</span>cluster), <span class="fu">unlist</span>(true_cluster))</span></code></pre></div>
<p>We provide two methods to correct the batch effects in gene
expression level. Method (1) is using only PRECAST results to obtain the
batch corrected gene expressions if the species of data is unknown or
the number of overlapped housekeeping genes between the variable genes
in <code>PRECASTObj@seulist</code> and the genes in database is less
than five. Method (2) is using bouth housekeeping gene and PRECAST
results to obtain the batch corrected gene expressions.</p>
<ul>
<li>Note: to obtain batch corrected gene expressions based on
housekeeping genes as the negative control, users must specify the
species of data source and use gene symbol names in
<code>PRECASTObj@seulist</code>.</li>
</ul>
<p>Integrate the two samples by the function
<code>IntegrateSpaData</code>. Because this is a simulated data, we use
Method (1) by setting <code>species=&#39;unknown&#39;</code>.</p>
<div class="sourceCode" id="cb14"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1" tabindex="-1"></a></span>
<span id="cb14-2"><a href="#cb14-2" tabindex="-1"></a>seuInt <span class="ot">&lt;-</span> <span class="fu">IntegrateSpaData</span>(PRECASTObj, <span class="at">species=</span><span class="st">&#39;unknown&#39;</span>)</span>
<span id="cb14-3"><a href="#cb14-3" tabindex="-1"></a>seuInt </span>
<span id="cb14-4"><a href="#cb14-4" tabindex="-1"></a><span class="do">## The low-dimensional embeddings obtained by PRECAST are saved in PRECAST reduction slot.</span></span></code></pre></div>
</div>
</div>
<div id="visualization" class="section level2">
<h2>Visualization</h2>
<p>First, user can choose a beautiful color schema using
<code>chooseColors()</code>.</p>
<div class="sourceCode" id="cb15"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1" tabindex="-1"></a>cols_cluster <span class="ot">&lt;-</span> <span class="fu">chooseColors</span>(<span class="at">palettes_name =</span> <span class="st">&#39;Nature 10&#39;</span>, <span class="at">n_colors =</span> <span class="dv">7</span>, <span class="at">plot_colors =</span> <span class="cn">TRUE</span>)</span></code></pre></div>
<p>Show the spatial scatter plot for clusters</p>
<div class="sourceCode" id="cb16"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb16-1"><a href="#cb16-1" tabindex="-1"></a>p12 <span class="ot">&lt;-</span> <span class="fu">SpaPlot</span>(seuInt, <span class="at">batch=</span><span class="cn">NULL</span>, <span class="at">cols=</span>cols_cluster, <span class="at">point_size=</span><span class="dv">2</span>, <span class="at">combine=</span><span class="cn">TRUE</span>)</span>
<span id="cb16-2"><a href="#cb16-2" tabindex="-1"></a>p12</span>
<span id="cb16-3"><a href="#cb16-3" tabindex="-1"></a><span class="co"># users can plot each sample by setting combine=FALSE</span></span></code></pre></div>
<p>Users can re-plot the above figures for specific need by returning a
ggplot list object. For example, we only plot the spatial heatmap of
first two data batches.</p>
<div class="sourceCode" id="cb17"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb17-1"><a href="#cb17-1" tabindex="-1"></a>pList <span class="ot">&lt;-</span> <span class="fu">SpaPlot</span>(seuInt, <span class="at">batch=</span><span class="cn">NULL</span>, <span class="at">cols=</span>cols_cluster, <span class="at">point_size=</span><span class="dv">2</span>, <span class="at">combine=</span><span class="cn">FALSE</span>, <span class="at">title_name=</span><span class="cn">NULL</span>)</span>
<span id="cb17-2"><a href="#cb17-2" tabindex="-1"></a><span class="fu">drawFigs</span>(pList[<span class="dv">1</span><span class="sc">:</span><span class="dv">2</span>], <span class="at">layout.dim =</span> <span class="fu">c</span>(<span class="dv">1</span>,<span class="dv">2</span>), <span class="at">common.legend =</span> <span class="cn">TRUE</span>, <span class="at">legend.position =</span> <span class="st">&#39;right&#39;</span>, <span class="at">align=</span><span class="st">&#39;hv&#39;</span>)</span></code></pre></div>
<p>Show the spatial UMAP/tNSE RGB plot</p>
<div class="sourceCode" id="cb18"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb18-1"><a href="#cb18-1" tabindex="-1"></a>seuInt <span class="ot">&lt;-</span> <span class="fu">AddUMAP</span>(seuInt) </span>
<span id="cb18-2"><a href="#cb18-2" tabindex="-1"></a><span class="fu">SpaPlot</span>(seuInt, <span class="at">batch=</span><span class="cn">NULL</span>,<span class="at">item=</span><span class="st">&#39;RGB_UMAP&#39;</span>,<span class="at">point_size=</span><span class="dv">1</span>, <span class="at">combine=</span><span class="cn">TRUE</span>, <span class="at">text_size=</span><span class="dv">15</span>)</span>
<span id="cb18-3"><a href="#cb18-3" tabindex="-1"></a></span>
<span id="cb18-4"><a href="#cb18-4" tabindex="-1"></a><span class="do">## Plot tSNE RGB plot</span></span>
<span id="cb18-5"><a href="#cb18-5" tabindex="-1"></a><span class="co">#seuInt &lt;- AddTSNE(seuInt) </span></span>
<span id="cb18-6"><a href="#cb18-6" tabindex="-1"></a><span class="co">#SpaPlot(seuInt, batch=NULL,item=&#39;RGB_TSNE&#39;,point_size=2, combine=T, text_size=15)</span></span></code></pre></div>
<p>Show the tSNE plot based on the extracted features from PRECAST to
check the performance of integration.</p>
<div class="sourceCode" id="cb19"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb19-1"><a href="#cb19-1" tabindex="-1"></a>seuInt <span class="ot">&lt;-</span> <span class="fu">AddTSNE</span>(seuInt, <span class="at">n_comp =</span> <span class="dv">2</span>) </span>
<span id="cb19-2"><a href="#cb19-2" tabindex="-1"></a></span>
<span id="cb19-3"><a href="#cb19-3" tabindex="-1"></a>p1 <span class="ot">&lt;-</span> <span class="fu">dimPlot</span>(seuInt, <span class="at">item=</span><span class="st">&#39;cluster&#39;</span>, <span class="at">font_family=</span><span class="st">&#39;serif&#39;</span>, <span class="at">cols=</span>cols_cluster) <span class="co"># Times New Roman</span></span>
<span id="cb19-4"><a href="#cb19-4" tabindex="-1"></a>p2 <span class="ot">&lt;-</span> <span class="fu">dimPlot</span>(seuInt, <span class="at">item=</span><span class="st">&#39;batch&#39;</span>, <span class="at">point_size =</span> <span class="dv">1</span>,  <span class="at">font_family=</span><span class="st">&#39;serif&#39;</span>)</span>
<span id="cb19-5"><a href="#cb19-5" tabindex="-1"></a><span class="fu">drawFigs</span>(<span class="fu">list</span>(p1, p2), <span class="at">common.legend=</span><span class="cn">FALSE</span>, <span class="at">align=</span><span class="st">&#39;hv&#39;</span>) </span>
<span id="cb19-6"><a href="#cb19-6" tabindex="-1"></a><span class="co"># It is noted that only sample batch 1 has cluster 4, and only sample batch 2 has cluster 7. </span></span></code></pre></div>
<p>Show the UMAP plot based on the extracted features from PRECAST.</p>
<div class="sourceCode" id="cb20"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb20-1"><a href="#cb20-1" tabindex="-1"></a><span class="fu">dimPlot</span>(seuInt, <span class="at">reduction =</span> <span class="st">&#39;UMAP3&#39;</span>, <span class="at">item=</span><span class="st">&#39;cluster&#39;</span>, <span class="at">cols=</span>cols_cluster, <span class="at">font_family=</span><span class="st">&#39;serif&#39;</span>)</span></code></pre></div>
<p>Users can also use the visualization functions in Seurat package:</p>
<div class="sourceCode" id="cb21"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb21-1"><a href="#cb21-1" tabindex="-1"></a><span class="fu">library</span>(Seurat)</span>
<span id="cb21-2"><a href="#cb21-2" tabindex="-1"></a>p1 <span class="ot">&lt;-</span> <span class="fu">DimPlot</span>(seuInt[,<span class="dv">1</span><span class="sc">:</span> <span class="dv">4226</span>], <span class="at">reduction =</span> <span class="st">&#39;position&#39;</span>, <span class="at">cols=</span>cols_cluster, <span class="at">pt.size =</span><span class="dv">1</span>) <span class="co"># plot the first data batch: first 4226 spots.</span></span>
<span id="cb21-3"><a href="#cb21-3" tabindex="-1"></a>p2 <span class="ot">&lt;-</span> <span class="fu">DimPlot</span>(seuInt, <span class="at">reduction =</span> <span class="st">&#39;tSNE&#39;</span>,<span class="at">cols=</span>cols_cluster, <span class="at">pt.size=</span><span class="dv">1</span>)</span>
<span id="cb21-4"><a href="#cb21-4" tabindex="-1"></a><span class="fu">drawFigs</span>(<span class="fu">list</span>(p1, p2), <span class="at">layout.dim =</span> <span class="fu">c</span>(<span class="dv">1</span>,<span class="dv">2</span>), <span class="at">common.legend =</span> <span class="cn">TRUE</span>)</span></code></pre></div>
<p>Combined differential expression analysis</p>
<div class="sourceCode" id="cb22"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb22-1"><a href="#cb22-1" tabindex="-1"></a>dat_deg <span class="ot">&lt;-</span> <span class="fu">FindAllMarkers</span>(seuInt)</span>
<span id="cb22-2"><a href="#cb22-2" tabindex="-1"></a><span class="fu">library</span>(dplyr)</span>
<span id="cb22-3"><a href="#cb22-3" tabindex="-1"></a>n <span class="ot">&lt;-</span> <span class="dv">2</span></span>
<span id="cb22-4"><a href="#cb22-4" tabindex="-1"></a>dat_deg <span class="sc">%&gt;%</span></span>
<span id="cb22-5"><a href="#cb22-5" tabindex="-1"></a>  <span class="fu">group_by</span>(cluster) <span class="sc">%&gt;%</span></span>
<span id="cb22-6"><a href="#cb22-6" tabindex="-1"></a>  <span class="fu">top_n</span>(<span class="at">n =</span> n, <span class="at">wt =</span> avg_log2FC) <span class="ot">-&gt;</span> top10</span>
<span id="cb22-7"><a href="#cb22-7" tabindex="-1"></a></span>
<span id="cb22-8"><a href="#cb22-8" tabindex="-1"></a><span class="fu">head</span>(top10)</span></code></pre></div>
<details>
<summary>
<strong>Session Info</strong>
</summary>
<div class="sourceCode" id="cb23"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb23-1"><a href="#cb23-1" tabindex="-1"></a><span class="fu">sessionInfo</span>()</span>
<span id="cb23-2"><a href="#cb23-2" tabindex="-1"></a><span class="co">#&gt; R version 4.4.1 (2024-06-14 ucrt)</span></span>
<span id="cb23-3"><a href="#cb23-3" tabindex="-1"></a><span class="co">#&gt; Platform: x86_64-w64-mingw32/x64</span></span>
<span id="cb23-4"><a href="#cb23-4" tabindex="-1"></a><span class="co">#&gt; Running under: Windows 11 x64 (build 26100)</span></span>
<span id="cb23-5"><a href="#cb23-5" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb23-6"><a href="#cb23-6" tabindex="-1"></a><span class="co">#&gt; Matrix products: default</span></span>
<span id="cb23-7"><a href="#cb23-7" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb23-8"><a href="#cb23-8" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb23-9"><a href="#cb23-9" tabindex="-1"></a><span class="co">#&gt; locale:</span></span>
<span id="cb23-10"><a href="#cb23-10" tabindex="-1"></a><span class="co">#&gt; [1] LC_COLLATE=C                               </span></span>
<span id="cb23-11"><a href="#cb23-11" tabindex="-1"></a><span class="co">#&gt; [2] LC_CTYPE=Chinese (Simplified)_China.utf8   </span></span>
<span id="cb23-12"><a href="#cb23-12" tabindex="-1"></a><span class="co">#&gt; [3] LC_MONETARY=Chinese (Simplified)_China.utf8</span></span>
<span id="cb23-13"><a href="#cb23-13" tabindex="-1"></a><span class="co">#&gt; [4] LC_NUMERIC=C                               </span></span>
<span id="cb23-14"><a href="#cb23-14" tabindex="-1"></a><span class="co">#&gt; [5] LC_TIME=Chinese (Simplified)_China.utf8    </span></span>
<span id="cb23-15"><a href="#cb23-15" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb23-16"><a href="#cb23-16" tabindex="-1"></a><span class="co">#&gt; time zone: Asia/Shanghai</span></span>
<span id="cb23-17"><a href="#cb23-17" tabindex="-1"></a><span class="co">#&gt; tzcode source: internal</span></span>
<span id="cb23-18"><a href="#cb23-18" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb23-19"><a href="#cb23-19" tabindex="-1"></a><span class="co">#&gt; attached base packages:</span></span>
<span id="cb23-20"><a href="#cb23-20" tabindex="-1"></a><span class="co">#&gt; [1] parallel  stats     graphics  grDevices utils     datasets  methods  </span></span>
<span id="cb23-21"><a href="#cb23-21" tabindex="-1"></a><span class="co">#&gt; [8] base     </span></span>
<span id="cb23-22"><a href="#cb23-22" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb23-23"><a href="#cb23-23" tabindex="-1"></a><span class="co">#&gt; other attached packages:</span></span>
<span id="cb23-24"><a href="#cb23-24" tabindex="-1"></a><span class="co">#&gt; [1] PRECAST_1.7  gtools_3.9.5</span></span>
<span id="cb23-25"><a href="#cb23-25" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb23-26"><a href="#cb23-26" tabindex="-1"></a><span class="co">#&gt; loaded via a namespace (and not attached):</span></span>
<span id="cb23-27"><a href="#cb23-27" tabindex="-1"></a><span class="co">#&gt;   [1] RcppAnnoy_0.0.22            splines_4.4.1              </span></span>
<span id="cb23-28"><a href="#cb23-28" tabindex="-1"></a><span class="co">#&gt;   [3] later_1.3.2                 tibble_3.2.1               </span></span>
<span id="cb23-29"><a href="#cb23-29" tabindex="-1"></a><span class="co">#&gt;   [5] polyclip_1.10-7             fastDummies_1.7.4          </span></span>
<span id="cb23-30"><a href="#cb23-30" tabindex="-1"></a><span class="co">#&gt;   [7] lifecycle_1.0.4             rstatix_0.7.2              </span></span>
<span id="cb23-31"><a href="#cb23-31" tabindex="-1"></a><span class="co">#&gt;   [9] globals_0.16.3              lattice_0.22-6             </span></span>
<span id="cb23-32"><a href="#cb23-32" tabindex="-1"></a><span class="co">#&gt;  [11] MASS_7.3-60.2               backports_1.5.0            </span></span>
<span id="cb23-33"><a href="#cb23-33" tabindex="-1"></a><span class="co">#&gt;  [13] magrittr_2.0.3              plotly_4.10.4              </span></span>
<span id="cb23-34"><a href="#cb23-34" tabindex="-1"></a><span class="co">#&gt;  [15] sass_0.4.9                  rmarkdown_2.28             </span></span>
<span id="cb23-35"><a href="#cb23-35" tabindex="-1"></a><span class="co">#&gt;  [17] jquerylib_0.1.4             yaml_2.3.10                </span></span>
<span id="cb23-36"><a href="#cb23-36" tabindex="-1"></a><span class="co">#&gt;  [19] httpuv_1.6.15               Seurat_5.1.0               </span></span>
<span id="cb23-37"><a href="#cb23-37" tabindex="-1"></a><span class="co">#&gt;  [21] sctransform_0.4.1           spam_2.10-0                </span></span>
<span id="cb23-38"><a href="#cb23-38" tabindex="-1"></a><span class="co">#&gt;  [23] sp_2.1-4                    spatstat.sparse_3.1-0      </span></span>
<span id="cb23-39"><a href="#cb23-39" tabindex="-1"></a><span class="co">#&gt;  [25] reticulate_1.39.0           cowplot_1.1.3              </span></span>
<span id="cb23-40"><a href="#cb23-40" tabindex="-1"></a><span class="co">#&gt;  [27] pbapply_1.7-2               RColorBrewer_1.1-3         </span></span>
<span id="cb23-41"><a href="#cb23-41" tabindex="-1"></a><span class="co">#&gt;  [29] abind_1.4-8                 zlibbioc_1.50.0            </span></span>
<span id="cb23-42"><a href="#cb23-42" tabindex="-1"></a><span class="co">#&gt;  [31] Rtsne_0.17                  GenomicRanges_1.56.2       </span></span>
<span id="cb23-43"><a href="#cb23-43" tabindex="-1"></a><span class="co">#&gt;  [33] purrr_1.0.2                 BiocGenerics_0.50.0        </span></span>
<span id="cb23-44"><a href="#cb23-44" tabindex="-1"></a><span class="co">#&gt;  [35] GenomeInfoDbData_1.2.12     IRanges_2.38.1             </span></span>
<span id="cb23-45"><a href="#cb23-45" tabindex="-1"></a><span class="co">#&gt;  [37] S4Vectors_0.42.1            ggrepel_0.9.6              </span></span>
<span id="cb23-46"><a href="#cb23-46" tabindex="-1"></a><span class="co">#&gt;  [39] irlba_2.3.5.1               listenv_0.9.1              </span></span>
<span id="cb23-47"><a href="#cb23-47" tabindex="-1"></a><span class="co">#&gt;  [41] spatstat.utils_3.1-0        goftest_1.2-3              </span></span>
<span id="cb23-48"><a href="#cb23-48" tabindex="-1"></a><span class="co">#&gt;  [43] RSpectra_0.16-2             spatstat.random_3.3-2      </span></span>
<span id="cb23-49"><a href="#cb23-49" tabindex="-1"></a><span class="co">#&gt;  [45] fitdistrplus_1.2-1          parallelly_1.38.0          </span></span>
<span id="cb23-50"><a href="#cb23-50" tabindex="-1"></a><span class="co">#&gt;  [47] DelayedMatrixStats_1.26.0   leiden_0.4.3.1             </span></span>
<span id="cb23-51"><a href="#cb23-51" tabindex="-1"></a><span class="co">#&gt;  [49] codetools_0.2-20            DelayedArray_0.30.1        </span></span>
<span id="cb23-52"><a href="#cb23-52" tabindex="-1"></a><span class="co">#&gt;  [51] scuttle_1.14.0              tidyselect_1.2.1           </span></span>
<span id="cb23-53"><a href="#cb23-53" tabindex="-1"></a><span class="co">#&gt;  [53] UCSC.utils_1.0.0            farver_2.1.2               </span></span>
<span id="cb23-54"><a href="#cb23-54" tabindex="-1"></a><span class="co">#&gt;  [55] ScaledMatrix_1.12.0         viridis_0.6.5              </span></span>
<span id="cb23-55"><a href="#cb23-55" tabindex="-1"></a><span class="co">#&gt;  [57] matrixStats_1.4.1           stats4_4.4.1               </span></span>
<span id="cb23-56"><a href="#cb23-56" tabindex="-1"></a><span class="co">#&gt;  [59] spatstat.explore_3.3-2      jsonlite_1.8.9             </span></span>
<span id="cb23-57"><a href="#cb23-57" tabindex="-1"></a><span class="co">#&gt;  [61] BiocNeighbors_1.22.0        Formula_1.2-5              </span></span>
<span id="cb23-58"><a href="#cb23-58" tabindex="-1"></a><span class="co">#&gt;  [63] progressr_0.14.0            ggridges_0.5.6             </span></span>
<span id="cb23-59"><a href="#cb23-59" tabindex="-1"></a><span class="co">#&gt;  [65] survival_3.6-4              scater_1.32.1              </span></span>
<span id="cb23-60"><a href="#cb23-60" tabindex="-1"></a><span class="co">#&gt;  [67] tools_4.4.1                 ica_1.0-3                  </span></span>
<span id="cb23-61"><a href="#cb23-61" tabindex="-1"></a><span class="co">#&gt;  [69] Rcpp_1.0.13                 glue_1.7.0                 </span></span>
<span id="cb23-62"><a href="#cb23-62" tabindex="-1"></a><span class="co">#&gt;  [71] gridExtra_2.3               SparseArray_1.4.8          </span></span>
<span id="cb23-63"><a href="#cb23-63" tabindex="-1"></a><span class="co">#&gt;  [73] xfun_0.47                   MatrixGenerics_1.16.0      </span></span>
<span id="cb23-64"><a href="#cb23-64" tabindex="-1"></a><span class="co">#&gt;  [75] ggthemes_5.1.0              GenomeInfoDb_1.40.1        </span></span>
<span id="cb23-65"><a href="#cb23-65" tabindex="-1"></a><span class="co">#&gt;  [77] dplyr_1.1.4                 formatR_1.14               </span></span>
<span id="cb23-66"><a href="#cb23-66" tabindex="-1"></a><span class="co">#&gt;  [79] fastmap_1.2.0               fansi_1.0.6                </span></span>
<span id="cb23-67"><a href="#cb23-67" tabindex="-1"></a><span class="co">#&gt;  [81] digest_0.6.37               rsvd_1.0.5                 </span></span>
<span id="cb23-68"><a href="#cb23-68" tabindex="-1"></a><span class="co">#&gt;  [83] R6_2.5.1                    mime_0.12                  </span></span>
<span id="cb23-69"><a href="#cb23-69" tabindex="-1"></a><span class="co">#&gt;  [85] colorspace_2.1-1            scattermore_1.2            </span></span>
<span id="cb23-70"><a href="#cb23-70" tabindex="-1"></a><span class="co">#&gt;  [87] tensor_1.5                  spatstat.data_3.1-2        </span></span>
<span id="cb23-71"><a href="#cb23-71" tabindex="-1"></a><span class="co">#&gt;  [89] utf8_1.2.4                  tidyr_1.3.1                </span></span>
<span id="cb23-72"><a href="#cb23-72" tabindex="-1"></a><span class="co">#&gt;  [91] generics_0.1.3              data.table_1.16.0          </span></span>
<span id="cb23-73"><a href="#cb23-73" tabindex="-1"></a><span class="co">#&gt;  [93] httr_1.4.7                  htmlwidgets_1.6.4          </span></span>
<span id="cb23-74"><a href="#cb23-74" tabindex="-1"></a><span class="co">#&gt;  [95] S4Arrays_1.4.1              uwot_0.2.2                 </span></span>
<span id="cb23-75"><a href="#cb23-75" tabindex="-1"></a><span class="co">#&gt;  [97] pkgconfig_2.0.3             gtable_0.3.5               </span></span>
<span id="cb23-76"><a href="#cb23-76" tabindex="-1"></a><span class="co">#&gt;  [99] lmtest_0.9-40               SingleCellExperiment_1.26.0</span></span>
<span id="cb23-77"><a href="#cb23-77" tabindex="-1"></a><span class="co">#&gt; [101] XVector_0.44.0              htmltools_0.5.8.1          </span></span>
<span id="cb23-78"><a href="#cb23-78" tabindex="-1"></a><span class="co">#&gt; [103] carData_3.0-5               dotCall64_1.1-1            </span></span>
<span id="cb23-79"><a href="#cb23-79" tabindex="-1"></a><span class="co">#&gt; [105] SeuratObject_5.0.2          scales_1.3.0               </span></span>
<span id="cb23-80"><a href="#cb23-80" tabindex="-1"></a><span class="co">#&gt; [107] Biobase_2.64.0              png_0.1-8                  </span></span>
<span id="cb23-81"><a href="#cb23-81" tabindex="-1"></a><span class="co">#&gt; [109] harmony_1.2.1               spatstat.univar_3.0-1      </span></span>
<span id="cb23-82"><a href="#cb23-82" tabindex="-1"></a><span class="co">#&gt; [111] knitr_1.48                  rstudioapi_0.16.0          </span></span>
<span id="cb23-83"><a href="#cb23-83" tabindex="-1"></a><span class="co">#&gt; [113] reshape2_1.4.4              nlme_3.1-164               </span></span>
<span id="cb23-84"><a href="#cb23-84" tabindex="-1"></a><span class="co">#&gt; [115] cachem_1.1.0                zoo_1.8-12                 </span></span>
<span id="cb23-85"><a href="#cb23-85" tabindex="-1"></a><span class="co">#&gt; [117] stringr_1.5.1               KernSmooth_2.23-24         </span></span>
<span id="cb23-86"><a href="#cb23-86" tabindex="-1"></a><span class="co">#&gt; [119] miniUI_0.1.1.1              vipor_0.4.7                </span></span>
<span id="cb23-87"><a href="#cb23-87" tabindex="-1"></a><span class="co">#&gt; [121] GiRaF_1.0.1                 pillar_1.9.0               </span></span>
<span id="cb23-88"><a href="#cb23-88" tabindex="-1"></a><span class="co">#&gt; [123] grid_4.4.1                  vctrs_0.6.5                </span></span>
<span id="cb23-89"><a href="#cb23-89" tabindex="-1"></a><span class="co">#&gt; [125] RANN_2.6.2                  promises_1.3.0             </span></span>
<span id="cb23-90"><a href="#cb23-90" tabindex="-1"></a><span class="co">#&gt; [127] ggpubr_0.6.0                car_3.1-3                  </span></span>
<span id="cb23-91"><a href="#cb23-91" tabindex="-1"></a><span class="co">#&gt; [129] BiocSingular_1.20.0         DR.SC_3.4                  </span></span>
<span id="cb23-92"><a href="#cb23-92" tabindex="-1"></a><span class="co">#&gt; [131] beachmat_2.20.0             xtable_1.8-4               </span></span>
<span id="cb23-93"><a href="#cb23-93" tabindex="-1"></a><span class="co">#&gt; [133] cluster_2.1.6               beeswarm_0.4.0             </span></span>
<span id="cb23-94"><a href="#cb23-94" tabindex="-1"></a><span class="co">#&gt; [135] evaluate_1.0.0              cli_3.6.3                  </span></span>
<span id="cb23-95"><a href="#cb23-95" tabindex="-1"></a><span class="co">#&gt; [137] compiler_4.4.1              rlang_1.1.4                </span></span>
<span id="cb23-96"><a href="#cb23-96" tabindex="-1"></a><span class="co">#&gt; [139] crayon_1.5.3                future.apply_1.11.2        </span></span>
<span id="cb23-97"><a href="#cb23-97" tabindex="-1"></a><span class="co">#&gt; [141] ggsignif_0.6.4              mclust_6.1.1               </span></span>
<span id="cb23-98"><a href="#cb23-98" tabindex="-1"></a><span class="co">#&gt; [143] plyr_1.8.9                  ggbeeswarm_0.7.2           </span></span>
<span id="cb23-99"><a href="#cb23-99" tabindex="-1"></a><span class="co">#&gt; [145] stringi_1.8.4               viridisLite_0.4.2          </span></span>
<span id="cb23-100"><a href="#cb23-100" tabindex="-1"></a><span class="co">#&gt; [147] deldir_2.0-4                BiocParallel_1.38.0        </span></span>
<span id="cb23-101"><a href="#cb23-101" tabindex="-1"></a><span class="co">#&gt; [149] munsell_0.5.1               lazyeval_0.2.2             </span></span>
<span id="cb23-102"><a href="#cb23-102" tabindex="-1"></a><span class="co">#&gt; [151] spatstat.geom_3.3-3         CompQuadForm_1.4.3         </span></span>
<span id="cb23-103"><a href="#cb23-103" tabindex="-1"></a><span class="co">#&gt; [153] Matrix_1.7-0                RcppHNSW_0.6.0             </span></span>
<span id="cb23-104"><a href="#cb23-104" tabindex="-1"></a><span class="co">#&gt; [155] patchwork_1.3.0             sparseMatrixStats_1.16.0   </span></span>
<span id="cb23-105"><a href="#cb23-105" tabindex="-1"></a><span class="co">#&gt; [157] future_1.34.0               ggplot2_3.5.2              </span></span>
<span id="cb23-106"><a href="#cb23-106" tabindex="-1"></a><span class="co">#&gt; [159] shiny_1.9.1                 SummarizedExperiment_1.34.0</span></span>
<span id="cb23-107"><a href="#cb23-107" tabindex="-1"></a><span class="co">#&gt; [161] ROCR_1.0-11                 broom_1.0.7                </span></span>
<span id="cb23-108"><a href="#cb23-108" tabindex="-1"></a><span class="co">#&gt; [163] igraph_2.0.3                bslib_0.8.0</span></span></code></pre></div>
</details>
</div>



<!-- code folding -->


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
